
---
title: "Bulk RNA Seq data analysis"
author: "Yekazci"
date: "03 09 2022"
output: html_document
---

### This Rmarkdown file contains the commands starting from exploratory and differential expression analysis with DESeq2 to
### visualisation of significant genes in spesific samples, identification and heatmaps of genes overrepresented in spesific
### GO terms etc.


 setwd("E:/TRANSCRIPTOME/1_Main_TRANSCRIPTOME/DRG_TRANSCRIPTOME_1_ANALYSIS/SALMON_DRG_NG_ANALYSIS_1")  ### directory contains the pseudo alignment files generated by SALMON.

 list.files()

 dir = getwd()

 salmon_drg_ng_samples = read.table(file.path(dir, "samples_drg_ng.txt"), header = TRUE)

 salmon_drg_ng_samples

 Samples

# 1 DRG_CS1_071217_S4_salmon_quant
# 2 DRG_CS2_071217_S5_salmon_quant
# 3 DRG_F_051017_S2_salmon_quant
# 4 DRG_F_121017_S1_salmon_quant
# 5 DRG_F_191017_S3_salmon_quant
# 6 NG_CS_071217_S8_salmon_quant
# 7 NG_CS_141217_S9_salmon_quant
# 8 NG_F_121017_S6_salmon_quant
# 9 G_F_191017_S7_salmon_quant

  sample_table_drg_ng = read.table("sample_table_drg_ng.txt", header = TRUE)

  sample_table_drg_ng

#   SampleNames     condition
# 1  DRG_CS1_S4      cont_drg
# 2  DRG_CS2_S5      cont_drg
# 3    DRG_F_S2    sample_drg
# 4    DRG_F_S1    sample_drg
# 5    DRG_F_S3    sample_drg
# 6    NG_CS_S8   cont_nodose
# 7    NG_CS_S9   cont_nodose
# 8     NG_F_S6 sample_nodose
# 9     NG_F_S7 sample_nodose

files <- file.path(dir,  "SALMON_DRG_RUN_1_QUANT_FILES", salmon_drg_ng_samples$Samples, "quant.sf")

   names(files) = paste0(sample_table_drg_ng$SampleNames)
 all(file.exists(files))

 # [1] TRUE

 files

# DRG_CS1_S4
# "E:/TRANSCRIPTOME/1_Main_TRANSCRIPTOME/DRG_TRANSCRIPTOME_1_ANALYSIS/SALMON_DRG_NG_ANALYSIS_1/SALMON_DRG_RUN_1_QUANT_FILES/DRG_CS1_071217_S4_salmon_quant/quant.sf"
# DRG_CS2_S5
# "E:/TRANSCRIPTOME/1_Main_TRANSCRIPTOME/DRG_TRANSCRIPTOME_1_ANALYSIS/SALMON_DRG_NG_ANALYSIS_1/SALMON_DRG_RUN_1_QUANT_FILES/DRG_CS2_071217_S5_salmon_quant/quant.sf"
# DRG_F_S2
# "E:/TRANSCRIPTOME/1_Main_TRANSCRIPTOME/DRG_TRANSCRIPTOME_1_ANALYSIS/SALMON_DRG_NG_ANALYSIS_1/SALMON_DRG_RUN_1_QUANT_FILES/DRG_F_051017_S2_salmon_quant/quant.sf"
# DRG_F_S1
# "E:/TRANSCRIPTOME/1_Main_TRANSCRIPTOME/DRG_TRANSCRIPTOME_1_ANALYSIS/SALMON_DRG_NG_ANALYSIS_1/SALMON_DRG_RUN_1_QUANT_FILES/DRG_F_121017_S1_salmon_quant/quant.sf"
# DRG_F_S3
# "E:/TRANSCRIPTOME/1_Main_TRANSCRIPTOME/DRG_TRANSCRIPTOME_1_ANALYSIS/SALMON_DRG_NG_ANALYSIS_1/SALMON_DRG_RUN_1_QUANT_FILES/DRG_F_191017_S3_salmon_quant/quant.sf"
# NG_CS_S8
# "E:/TRANSCRIPTOME/1_Main_TRANSCRIPTOME/DRG_TRANSCRIPTOME_1_ANALYSIS/SALMON_DRG_NG_ANALYSIS_1/SALMON_DRG_RUN_1_QUANT_FILES/NG_CS_071217_S8_salmon_quant/quant.sf"
# NG_CS_S9
# "E:/TRANSCRIPTOME/1_Main_TRANSCRIPTOME/DRG_TRANSCRIPTOME_1_ANALYSIS/SALMON_DRG_NG_ANALYSIS_1/SALMON_DRG_RUN_1_QUANT_FILES/NG_CS_141217_S9_salmon_quant/quant.sf"
# NG_F_S6
# "E:/TRANSCRIPTOME/1_Main_TRANSCRIPTOME/DRG_TRANSCRIPTOME_1_ANALYSIS/SALMON_DRG_NG_ANALYSIS_1/SALMON_DRG_RUN_1_QUANT_FILES/NG_F_121017_S6_salmon_quant/quant.sf"
# NG_F_S7
# "E:/TRANSCRIPTOME/1_Main_TRANSCRIPTOME/DRG_TRANSCRIPTOME_1_ANALYSIS/SALMON_DRG_NG_ANALYSIS_1/SALMON_DRG_RUN_1_QUANT_FILES/NG_F_191017_S7_salmon_quant/quant.sf"

 source("https://bioconductor.org/biocLite.R")

  biocLite("EnsDb.Mmusculus.v79")

#### different versions will be compared for tx2gene dataframes.

 ### make our own txdb for ensembl v91:

   library(ensembldb)

mouse_txdb_ens_v91 <- makeTxDbFromGFF("Mus_musculus.GRCm38.91.gtf/Mus_musculus.GRCm38.91.gtf",format="gtf")

# Import genomic features from the file as a GRanges object ... OK
# Prepare the 'metadata' data frame ... OK
# Make the TxDb object ... OK
# Warning message:
#   In .get_cds_IDX(type, phase) :
#   The "phase" metadata column contains non-NA values for features of type stop_codon. This information was ignored.

 keys_mouse_ens_v91 = keys(mouse_txdb_ens_v91, keytype = "GENEID")
 df_mouse_ens_91 <- select(mouse_txdb_ens_v91, keys = keys_mouse_ens_v91, keytype = "GENEID", columns = "TXNAME")

 # 'select()' returned 1:many mapping between keys and columns

 head(df_mouse_ens_91)

#  GENEID             TXNAME
# 1 ENSMUSG00000000001 ENSMUST00000000001
# 2 ENSMUSG00000000003 ENSMUST00000000003
# 3 ENSMUSG00000000003 ENSMUST00000114041
# 4 ENSMUSG00000000028 ENSMUST00000000028
# 5 ENSMUSG00000000028 ENSMUST00000096990
# 6 ENSMUSG00000000028 ENSMUST00000115585
 tx2gene_mouse_ens_91 = df_mouse_ens_91[, 2:1]  # tx ID, then gene ID
 head(tx2gene_mouse_ens_91)

#  TXNAME             GENEID
# 1 ENSMUST00000000001 ENSMUSG00000000001
# 2 ENSMUST00000000003 ENSMUSG00000000003
# 3 ENSMUST00000114041 ENSMUSG00000000003
# 4 ENSMUST00000000028 ENSMUSG00000000028
# 5 ENSMUST00000096990 ENSMUSG00000000028
# 6 ENSMUST00000115585 ENSMUSG00000000028

  length(tx2gene_mouse_ens_91$TXNAME)

  # [1] 133944

##### Other versions:

 library(TxDb.Mmusculus.UCSC.mm10.ensGene)

  keys_mouse_UCSC.mm10.ensGene = keys(TxDb.Mmusculus.UCSC.mm10.ensGene, keytype = "GENEID")

  df_mouse_UCSC.mm10.ensGene <- select(TxDb.Mmusculus.UCSC.mm10.ensGene, keys = keys_mouse_UCSC.mm10.ensGene, keytype = "GENEID", columns = "TXNAME")

  # 'select()' returned 1:many mapping between keys and columns

   head(df_mouse_UCSC.mm10.ensGene)

#    GENEID             TXNAME
# 1 ENSMUSG00000000001 ENSMUST00000000001
# 2 ENSMUSG00000000003 ENSMUST00000000003
# 3 ENSMUSG00000000003 ENSMUST00000114041
# 4 ENSMUSG00000000028 ENSMUST00000000028
# 5 ENSMUSG00000000028 ENSMUST00000096990
# 6 ENSMUSG00000000028 ENSMUST00000115585

 tx2gene_mouse_UCSC.mm10.ensGene = df_mouse_UCSC.mm10.ensGene[, 2:1]  # tx ID, then gene ID
 head(tx2gene_mouse_UCSC.mm10.ensGene)

#  TXNAME             GENEID
# 1 ENSMUST00000000001 ENSMUSG00000000001
# 2 ENSMUST00000000003 ENSMUSG00000000003
# 3 ENSMUST00000114041 ENSMUSG00000000003
# 4 ENSMUST00000000028 ENSMUSG00000000028
# 5 ENSMUST00000096990 ENSMUSG00000000028
# 6 ENSMUST00000115585 ENSMUSG00000000028

 length(tx2gene_mouse_UCSC.mm10.ensGene$TXNAME)

 # [1] 94647

  ### other version:

   library(EnsDb.Mmusculus.v79)
 txdb_mouse_ens_v79 = EnsDb.Mmusculus.v79
 keys_mouse_ens_v79 = keys(txdb_mouse_ens_v79, keytype = "GENEID")

   df_mouse_ens_v79 <- select(txdb_mouse_ens_v79, keys = keys_mouse_ens_v79, keytype = "GENEID", columns = "TXNAME")

   head(df_mouse_ens_v79)

#    GENEID             TXNAME
# 1 ENSMUSG00000000001 ENSMUST00000000001
# 2 ENSMUSG00000000003 ENSMUST00000000003
# 3 ENSMUSG00000000003 ENSMUST00000114041
# 4 ENSMUSG00000000028 ENSMUST00000000028
# 5 ENSMUSG00000000028 ENSMUST00000096990
# 6 ENSMUSG00000000028 ENSMUST00000115585

   tx2gene_mouse_ens_v79 = df_mouse_ens_v79[, 2:1]  # tx ID, then gene ID

   head(tx2gene_mouse_ens_v79)

#    TXNAME             GENEID
# 1 ENSMUST00000000001 ENSMUSG00000000001
# 2 ENSMUST00000000003 ENSMUSG00000000003
# 3 ENSMUST00000114041 ENSMUSG00000000003
# 4 ENSMUST00000000028 ENSMUSG00000000028
# 5 ENSMUST00000096990 ENSMUSG00000000028
# 6 ENSMUST00000115585 ENSMUSG00000000028

 length(tx2gene_mouse_ens_v79$TXNAME)

 # [1] 104129

 intersect_of_versions = intersect(tx2gene_mouse_ens_91$TXNAME, tx2gene_mouse_ens_v79$TXNAME)
 length(intersect_of_versions)

 # [1] 101977

 intersect_of_versions_2 = intersect(tx2gene_mouse_ens_91$TXNAME, tx2gene_mouse_UCSC.mm10.ensGene$TXNAME)
 length(intersect_of_versions_2)

 # [1] 91659

  #### Last version ens.v91 containes most transcript IDs in other previous versions.

   library(tximport)
 txi_mouse_ens_91_drg_nod = tximport(files, type = "salmon",ignoreTxVersion = TRUE ,tx2gene = tx2gene_mouse_ens_91)

#  reading in files with read_tsv
# 1 2 3 4 5 6 7 8 9
# transcripts missing from tx2gene: 74
# summarizing abundance
# summarizing counts
# summarizing length

   txi_mouse_ens_v79_drg_nod = tximport(files, type = "salmon",ignoreTxVersion = TRUE ,tx2gene = tx2gene_mouse_ens_v79)

#    reading in files with read_tsv
# 1 2 3 4 5 6 7 8 9
# transcripts missing from tx2gene: 24212
# summarizing abundance
# summarizing counts
# summarizing length

   txi_mouse_UCSC.mm10.ensGene_drg_nod = tximport(files, type = "salmon",ignoreTxVersion = TRUE ,tx2gene = tx2gene_mouse_UCSC.mm10.ensGene)

#    reading in files with read_tsv
# 1 2 3 4 5 6 7 8 9
# transcripts missing from tx2gene: 30599
# summarizing abundance
# summarizing counts
# summarizing length

   #### Number of missing transcripts are minimum in ensembl V.91 so we will go with it.

   ### Differential expression analysis with DESeq2:

    library(DESeq2)

    dds_drg_nod = DESeqDataSetFromTximport(txi_mouse_ens_91_drg_nod,sample_table_drg_ng, ~condition)

    nrow(dds_drg_nod)
#   [1] 34517

    dds_drg_nod = dds_drg_nod[rowSums(counts(dds_drg_nod)),  1,]

    nrow(dds_drg_nod)

#   [1] 25202

    rld_drg_nod = rlog(dds_drg_nod, blind = FALSE)

#   using 'avgTxLength' from assays(dds), correcting for library size

    rld_drg_nod_matrix = assay(rld_drg_nod)

        ### Exploratory analysis:

          ### Sample distances. Default function is euclidean.

          sampleDists_drg_nod = dist(t(rld_drg_nod_matrix))

    library("pheatmap")
    library(RColorBrewer)

    colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
    pheatmap(as.matrix(sampleDists_drg_nod), clustering_distance_rows = sampleDists_drg_nod, clustering_distance_cols = sampleDists_drg_nod, col = colors)

    plotPCA(rld_drg_nod, intgroup = c("condition"))
    pca_DATA = plotPCA(rld_drg_nod, intgroup = c("condition"), returnData=TRUE)

      pca_DATA
#                   PC1        PC2         group     condition       name
#   DRG_CS1_S4 -21.53953 -15.011176      cont_drg      cont_drg DRG_CS1_S4
#   DRG_CS2_S5 -21.00059 -16.976613      cont_drg      cont_drg DRG_CS2_S5
#   DRG_F_S2   -27.60085   9.043892    sample_drg    sample_drg   DRG_F_S2
#   DRG_F_S1   -26.58757   9.956530    sample_drg    sample_drg   DRG_F_S1
#   DRG_F_S3   -25.66674  10.399539    sample_drg    sample_drg   DRG_F_S3
#   NG_CS_S8    27.14208 -11.056663   cont_nodose   cont_nodose   NG_CS_S8
#   NG_CS_S9    33.31522  -9.122232   cont_nodose   cont_nodose   NG_CS_S9
#   NG_F_S6     30.44165  11.045561 sample_nodose sample_nodose    NG_F_S6
#   NG_F_S7     31.49633  11.721162 sample_nodose sample_nodose    NG_F_S7
    percentVar <- round(100 * attr(pca_DATA, "percentVar"))
    percentVar

#   [1] 74 14


    library("ggplot2")
    ggplot(pca_DATA, aes(PC1, PC2, color=name, shape=condition)) + geom_point(size=3) + xlab(paste0("PC1: ",percentVar[1],"% variance")) + ylab(paste0("PC2: ",percentVar[2],"% variance")) + coord_fixed()

      #############
    ### PCA 3D
      ntop = 500
    rv = rowVars(rld_drg_nod_matrix)
    selected = order(rv, decreasing = TRUE)[seq_len(min(ntop, length(rv)))]
    mat = t(rld_drg_nod_matrix[selected,])
    pca = prcomp(mat)
    pca$x
   # PC1        PC2       PC3        PC4        PC5        PC6
   # DRG_CS1_S4 -21.53953 -15.011176  0.219651 12.0798574 -2.3848644  1.2167092
   # DRG_CS2_S5 -21.00059 -16.976613  7.865052 -6.0799344  5.3944096 -3.7112809
   # DRG_F_S2   -27.60085   9.043892 -2.293106 -4.1221805 -8.3499010 -5.5141667
   # DRG_F_S1   -26.58757   9.956530 -8.664642  0.8728971  5.8833519 -0.1304913
   # DRG_F_S3   -25.66674  10.399539  3.598263 -2.2984627  0.3022581  8.2017283
   # NG_CS_S8    27.14208 -11.056663 -1.785396 -2.3301370 -5.2499460  1.2673076
   # NG_CS_S9    33.31522  -9.122232 -8.392474 -4.1407060  1.8045415  2.5471058
   # NG_F_S6     30.44165  11.045561  1.017519  5.1407865  3.6834515 -5.6590158
   # NG_F_S7     31.49633  11.721162  8.435134  0.8778796 -1.0833014  1.7821038
   # PC7         PC8          PC9
   # DRG_CS1_S4  1.8408755  0.04678494 1.017541e-14
   # DRG_CS2_S5  0.0649278  0.21713152 2.048946e-14
   # DRG_F_S2    2.3538709 -0.50055285 7.313748e-15
   # DRG_F_S1   -2.2268115  3.47785637 5.312656e-15
   # DRG_F_S3   -1.1880013 -3.30216495 8.041131e-15
   # NG_CS_S8   -7.1438964  0.91192815 6.376205e-15
   # NG_CS_S9    5.3796551 -1.25932403 7.589220e-15
   # NG_F_S6    -1.9011130 -3.60836870 2.166260e-15
   # NG_F_S7     2.8204928  4.01670956 4.011379e-15
    percentVar_pca3d <- pca$sdev^2 / sum( pca$sdev^2 )
    percentVar_pca3d = round(100 * percentVar_pca3d)
    percentVar_pca3d
   # [1] 74 14  3  3  2  2  1  1  0
    library(pca3d)
    pca3d(pca, components = 1:3, axes.color = "grey", bg = "white", radius = 1, show.labels = TRUE, axe.titles = c(paste0("PC1: %", percentVar_pca3d[1]),paste0("PC2: %", percentVar_pca3d[2]),paste0("PC3: %", percentVar_pca3d[3])) ,show.axe.titles = TRUE)
   # [1] 0.6663044 0.3395323 0.1732928
   # Creating new device

   ### Better labels:

    pca3d(pca, components = 1:3, axes.color = "grey", bg = "white", radius = 1, show.labels = TRUE, axe.titles = c(paste0("PC1: %", percentVar_pca3d[1]),paste0("PC2: %", percentVar_pca3d[2]),paste0("PC3: %", percentVar_pca3d[3])) ,show.axe.titles = TRUE, col = c("blue","blue", "yellow","yellow", "yellow", "green", "green", "red", "red"))
   # [1] 0.6663044 0.3395323 0.1732928
   # Creating new device

     ### Differential expression analysis:
       dds_drg_nod = DESeq(dds_drg_nod)


     res_DRG_F_vs_cont_drg = results(dds_drg_nod, contrast = c("condition", "sample_drg","cont_drg"), alpha = 0.05)

     res_NODS_F_vs_cont_nodose = results(dds_drg_nod, contrast = c("condition", "sample_nodose","cont_nodose"), alpha = 0.05)




       biocLite("org.Mm.eg.db")

     #### Filterings:
       res_DRG_F_vs_cont_drg_na_omit = na.omit(res_DRG_F_vs_cont_drg)
     res_NODS_F_vs_cont_nodose_na_omit = na.omit(res_NODS_F_vs_cont_nodose)

    ### OPTIONAL: "anyNA" function can be used as follows:

    anyNA(res_DRG_F_vs_cont_drg_na_omit, recursive=FALSE) ### It should return "FALSE".

     library(org.Mm.eg.db)

     ## Addition of gene symbols:
       res_DRG_F_vs_cont_drg_na_omit$symbols = mapIds(org.Mm.eg.db, keys = row.names(res_DRG_F_vs_cont_drg_na_omit), keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first" )


       res_NODS_F_vs_cont_nodose_na_omit$symbols = mapIds(org.Mm.eg.db, keys = row.names(res_NODS_F_vs_cont_nodose_na_omit), keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first" )


       table(duplicated(res_DRG_F_vs_cont_drg_na_omit$symbols))

    # FALSE  TRUE
    # 15854  2013

     table(duplicated(res_NODS_F_vs_cont_nodose_na_omit$symbols))

    # FALSE  TRUE
    # 15854  2013

     res_DRG_F_vs_cont_drg_na_omit


     ### We will order by padj values:
       res_DRG_F_vs_cont_drg_na_omit = res_DRG_F_vs_cont_drg_na_omit[order( res_DRG_F_vs_cont_drg_na_omit$padj),]
     res_NODS_F_vs_cont_nodose_na_omit = res_NODS_F_vs_cont_nodose_na_omit[order( res_NODS_F_vs_cont_nodose_na_omit$padj),]


     ### Genes without any symbol, usually pseudogenes, will be removed.
       res_DRG_F_vs_cont_drg_na_omit_symbols = na.omit(res_DRG_F_vs_cont_drg_na_omit)
     res_NODS_F_vs_cont_nodose_na_omit_symbols = na.omit(res_NODS_F_vs_cont_nodose_na_omit)

     res_DRG_F_vs_cont_drg_na_omit_symbols

     res_NODS_F_vs_cont_nodose_na_omit_symbols


     ### We will remove duplicated genes, only keeping the ones with lowest padj values.

       res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup = res_DRG_F_vs_cont_drg_na_omit_symbols[!duplicated(res_DRG_F_vs_cont_drg_na_omit_symbols$symbols),]
     res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup = res_NODS_F_vs_cont_nodose_na_omit_symbols[!duplicated(res_NODS_F_vs_cont_nodose_na_omit_symbols$symbols),]
     table(duplicated(res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup$symbols))

    #################################################################################

#     ### Bu komut sonucunda aslında "NA" diye isimlendirilen sembollerden en significant olanı hariç
#     diğerlerini çıkarmış olduk. Onlar arasında significant olanlar olabilir. Bu genlerin  bazılarının
#     sembolü var aslında, başka veritablarında. O yüzden sembol ekledikten sonra bu genlerin silinmesini engellemek için
#     ,sembol yerine "NA" içerenlerin sembol bölümüne ensembl id'leri yazılacak. Daha sonra duplicated
#     olanlar atılacak.

#################################################################################

# FALSE
# 15853

 table(duplicated(res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup$symbols))

# FALSE
# 15853

#### OPTIONAL: table(is.na(results_x$symbols) , bu komut "NA" içeren satırları veriyor.


 #### These lists will be used as backgrounds in GO analyzes.

 ### ENTREZ ids are added:
 res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup$entrez_IDs = mapIds(org.Mm.eg.db, keys = row.names(res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup), keytype = "ENSEMBL", column = "ENTREZID", multiVals = "first" )

 res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup$entrez_IDs = mapIds(org.Mm.eg.db, keys = row.names(res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup), keytype = "ENSEMBL", column = "ENTREZID", multiVals = "first" )

 res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup

 res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup

### GLOBAL HEATMAP:

 row_names_for_global_heatmap = intersect(row.names(res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup),row.names(res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup))
 length(row_names_for_global_heatmap)
# [1] 15842

 library(gplots)

 heatmap.2(rld_drg_nod_matrix[row_names_for_global_heatmap[1:12000],], scale="row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = NA, dendrogram = "col", cexCol = 0.70)

 ### GENE definitions
 res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup$gene_name = mapIds(org.Mm.eg.db, keys = row.names(res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup), keytype = "ENSEMBL", column = "GENENAME", multiVals = "first" )

 res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup$gene_name = mapIds(org.Mm.eg.db, keys = row.names(res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup), keytype = "ENSEMBL", column = "GENENAME", multiVals = "first" )

 res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup

 res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup

 ### Significant lists:

 ### Significant lists:
 res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup_sig = subset(res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup, padj < 0.05)
 res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup_sig = subset(res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup, padj < 0.05)
 ####
 ### LogFC  1.5 and LogFC < -1.5
 res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup_sig_LFC = subset(res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup_sig, abs(log2FoldChange),  1.5)
 res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup_sig_LFC = subset(res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup_sig, abs(log2FoldChange),  1.5)
 ####
 res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup_all_sig = res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup_sig
 res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup_all_sig =  res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup_sig
 ####
 ###Significant up and down separate lists:
 ### Up lists:

 res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup_sig_up_LFC_1.5 = subset(res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup_all_sig , (log2FoldChange),  1.5)
 res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup_sig_down_LFC_1.5 = subset(res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup_all_sig , (log2FoldChange) < -1.5)
 res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup_down_LFC_1.5 = subset(res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup_all_sig  , (log2FoldChange) < -1.5)
 res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup_up_LFC_1.5 = subset(res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup_all_sig  , (log2FoldChange),  1.5)
 ####

 ### EXPORT:
 write.csv(res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup, file = "res_DRG_F_vs_cont_drg_background.txt")

 write.csv(res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup_all_sig, file = "res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup_all_sig.txt")

 write.csv(res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup_sig_LFC, file = "res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup_sig_LFC.txt")

 write.csv(res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup_sig_up_LFC_1.5, file = "res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup_sig_up_LFC_1.5.txt")

 write.csv(res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup_sig_down_LFC_1.5, file = "res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup_sig_down_LFC_1.5.txt")

 ###

 write.csv(res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup, file = "res_NODS_F_vs_cont_nodose_background.txt")

 write.csv(res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup_all_sig, file = "res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup_all_sig.txt")

 write.csv(res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup_sig_LFC, file = "res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup_sig_LFC.txt")

 write.csv(res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup_down_LFC_1.5, file = "res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup_down_LFC_1.5.txt")

 write.csv( res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup_up_LFC_1.5, file = " res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup_up_LFC_1.5.txt")

 ####




#### EXTRA:

#### NOT: Aşağıda önceden Log2Fold Change belirlemenin analiz sonuçlarına etkisine bakıldı.

 res_DRG_F_vs_cont_drg_ifthreshold_1.5 = results(dds_drg_nod, contrast = c("condition", "sample_drg","cont_drg"), lfcThreshold=1.5)

 table(res_DRG_F_vs_cont_drg$pvalue < 0.05)

#FALSE  TRUE
#22233  2963

 table(res_DRG_F_vs_cont_drg_ifthreshold_1.5$pvalue < 0.05)

#FALSE  TRUE
#24958   238


 ### From deseq2 paper:
 ######"here we perform a statistical test for
# whether the log2 fold change is not greater than the threshold,
# as reflected in new p values."

#### Comparison of two methods:

 res_DRG_F_vs_cont_drg_try1 = subset(res_DRG_F_vs_cont_drg, abs(res_DRG_F_vs_cont_drg$log2FoldChange),  1.5)

#### NOT: Yukarıda absolute yerine sadece log2foldchange dememiz gerekiyordu belki de. Çünkü ifthreshold = +1.5 için karşılaştırma yaptık.
### Aşağıdaki değerlendirmenin doğruluğu net değil. Fakat her halükarda iki yöntem farklı sonuç veriyor. Bu kesin bilgi.

 table(res_DRG_F_vs_cont_drg_try1$pvalue < 0.05)

#FALSE  TRUE
# 2798   676

 ### Önce p-value threshold belirleyip sonra log2FC'ye göre değerlendirmek ile
     ### logFC değerini teste dahil etmek farklı sonuçlar veriyor.
       ### Makalede açıklandığı gibi ikinci analizde belli bir Log2FC değerine
       ### göre istatistiksel bir analiz gerçekleştiriyoruz.

      ### İkinci analizde bulduğumuz az sayıda gen,
      #birinci analizde bulduğumuz daha geniş listede de
      #tamamen içerilmektedir:

     res_DRG_F_vs_cont_drg_ifthreshold_1.5_x = subset(res_DRG_F_vs_cont_drg_ifthreshold_1.5  , pvalue < 0.05)
     res_DRG_F_vs_cont_drg_try1_x = subset(res_DRG_F_vs_cont_drg_try1  , pvalue < 0.05)
     xxx = intersect(row.names(res_DRG_F_vs_cont_drg_try1_x), row.names(res_DRG_F_vs_cont_drg_ifthreshold_1.5_x))
     length(xxx)
    # [1] 238  #### aynı genler, ifthreshold yöntemi ile bulunan)

      fpkm_DRG_NODOSE_run_1 = fpkm(dds_drg_nod)
      class(fpkm_DRG_NODOSE_run_1)
#     [1] "matrix"
      fpkm_DRG_NODOSE_run_1 = as.data.frame(fpkm_DRG_NODOSE_run_1)

     ### "fpkm()" function used avgTxLength of dds object for normalization:


      rld_drg_nod_DF = as.data.frame(rld_drg_nod_matrix)
      library(org.Mm.eg.db)

      rld_drg_nod_DF$symbols = mapIds(org.Mm.eg.db, keys = row.names(rld_drg_nod_DF), keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first" )
      rld_drg_nod_DF$GENE_NAME = mapIds(org.Mm.eg.db, keys = row.names(rld_drg_nod_DF), keytype = "ENSEMBL", column = "GENENAME", multiVals = "first" )


      length(rld_drg_nod_DF$DRG_CS1_S4)
#     [1] 25202
      length(fpkm_DRG_NODOSE_run_1$DRG_CS1_S4)
#     [1] 25202

      setdiff(row.names(rld_drg_nod_DF), row.names(fpkm_DRG_NODOSE_run_1))
#     character(0)

     ### they are same.

     #### we will take symbols and gene definitons from regularized log transformed data:
      fpkm_DRG_NODOSE_run_1$symbols = rld_drg_nod_DF$symbols
      fpkm_DRG_NODOSE_run_1$gene_name = rld_drg_nod_DF$GENE_NAME
      head(fpkm_DRG_NODOSE_run_1)

      ### EXPORT:
        write.csv(fpkm_DRG_NODOSE_run_1, file = "fpkm_DRG_NODOSE_run_1.txt")

       #### PC Component Genes:
        ### Percentage of contribution of genes to PC components:ü
        ## We use "pca" object obtained by prcomp:
        head(pca$rotation,5)

      aload = abs(pca$rotation)
      my_pca_genes = sweep(aload, 2, colSums(aload), "/")


      my_pca_genes_DF = as.data.frame(my_pca_genes)

      colSums(my_pca_genes_DF[,1:9])

      #     PC1 PC2 PC3 PC4 PC5 PC6 PC7 PC8 PC9

      #     1   1   1   1   1   1   1   1   1

      my_pca_genes_DF$symbols = mapIds(org.Mm.eg.db, keys=row.names(my_pca_genes_DF), column="SYMBOL", keytype="ENSEMBL", multiVals="first")

      my_pca_genes_DF$gene_name = mapIds(org.Mm.eg.db, keys=row.names(my_pca_genes_DF), column="GENENAME", keytype="ENSEMBL", multiVals="first")

      head(my_pca_genes_DF,5)
      write.csv(my_pca_genes_DF, file = "my_pca_genes_DF.txt")

     ##########################################################################################
     ##########################################################################################
     ##########################################################################################

     #### Heatmap'te kullanmak için duplicate sembollerin
     #### silindiği bir liste oluşturacaz:

     res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup

     res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup


     ### İkisinin uzunluğu da aynı. Şimdi farklarına bakacaz.

      setdiff(row.names(res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup), row.names(res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup))
#     [1] "ENSMUSG00000052305" "ENSMUSG00000078897" "ENSMUSG00000090093" "ENSMUSG00000090137"
#     [5] "ENSMUSG00000110195" "ENSMUSG00000110105" "ENSMUSG00000074513" "ENSMUSG00000029456"
#     [9] "ENSMUSG00000075569" "ENSMUSG00000048967" "ENSMUSG00000111410"
      setdiff(row.names(res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup),row.names(res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup))
#     [1] "ENSMUSG00000073940" "ENSMUSG00000036199" "ENSMUSG00000079109" "ENSMUSG00000030653"
#     [5] "ENSMUSG00000042647" "ENSMUSG00000102805" "ENSMUSG00000078880" "ENSMUSG00000068240"
#     [9] "ENSMUSG00000095730" "ENSMUSG00000078903" "ENSMUSG00000028698"


     ### Farklı olan genlerin hangileri olduğunu görmek için rld data frame'de bu genlere bakıyoruz.

      rld_drg_nod_DF[setdiff(row.names(res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup), row.names(res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup)),]


      rld_drg_nod_DF[setdiff(row.names(res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup),row.names(res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup)),]

       ### İkisi de aynı gen sembollerine sahip fakat gen ensembl no.ları farklı.

       ### DRG'nin tüm ensembl id'lerini alıp, nodose'da farklı olanları da ekleyerek
       #### tüm genleri içeren bir liste yapabiliriz:
       ### Nodozda farklı olanları depoladığımız geçici liste:

      template_genes = setdiff(row.names(res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup),row.names(res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup))
      template_genes
#     [1] "ENSMUSG00000073940" "ENSMUSG00000036199" "ENSMUSG00000079109" "ENSMUSG00000030653"
#     [5] "ENSMUSG00000042647" "ENSMUSG00000102805" "ENSMUSG00000078880" "ENSMUSG00000068240"
#     [9] "ENSMUSG00000095730" "ENSMUSG00000078903" "ENSMUSG00000028698"

     ### Birleştirme:

      all_genes_for_heatmaps = c(all_genes_for_heatmaps, template_genes)


      length(all_genes_for_heatmaps)
#     [1] 15864

     ### İki gruptaki tüm genleri içerip içermediğini test-etmek için :

      setdiff(row.names(res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup), all_genes_for_heatmaps)
#     character(0)
      setdiff(row.names(res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup),all_genes_for_heatmaps)
     character(0)

     ### Bu liste ensembl id'leri kullandığımız tüm heatmap'lerde kullanılacak.

     ### Nodoz ve DRG'nin özel heatmap'lerinde (Top Diff gibi) aşağıdaki listeleri
     ### RLD data-frame'e verecez.

      all_genes_DRG_no_duplicate = row.names(res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup)
      all_genes_NODOSE_no_duplicate = row.names(res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup)
      ####

       ### DRG ve Nodoz results listelerinde bulunan tüm genleri içeren rld df (data.frame) ve matriks:

        ####
        rld_drg_nod_df_filtered = rld_drg_nod_DF[all_genes_for_heatmaps,]
      rld_drg_nod_df_filtered_matrix = as.matrix(rld_drg_nod_df_filtered)
      ###

       ### HEATMAP label olarak kullanmak için:

        my_heatmap_labels = rld_drg_nod_df_filtered
      my_heatmap_labels[,1:10] = NULL ### count içeren sütunlar silindi. Sadece gen isimleri kaldı.
      head(my_heatmap_labels)


      #### PC1 ve PC2 genes:

      head(my_pca_genes_DF,1)

      PC1_genes_ordered = my_pca_genes_DF[order(my_pca_genes_DF$PC1,decreasing = TRUE),]

      PC2_genes_ordered = my_pca_genes_DF[order(my_pca_genes_DF$PC2,decreasing = TRUE),]

       ### To obtain ordered  PC1 and PC2 genes in data.frames , separately:

      PC1_genes_ordered <- subset( PC1_genes_ordered, select = c(PC1, symbols) )

      PC2_genes_ordered <- subset( PC2_genes_ordered, select = c(PC2, symbols) )

      write.csv(PC1_genes_ordered, file = "PC1_genes_ordered.txt")

      write.csv(PC2_genes_ordered, file = "PC2_genes_ordered.txt")

      ### FPKM matrices:

      fpkm_DRG_NODOSE_filtered_df = fpkm_DRG_NODOSE_run_1[all_genes_for_heatmaps,]
      head(fpkm_DRG_NODOSE_filtered_df)
      length(fpkm_DRG_NODOSE_filtered_df$DRG_CS1_S4)
#      [1] 15864
      length(fpkm_DRG_NODOSE_run_1$DRG_CS1_S4)
#      [1] 25202
      ### EXPORT filtered fpkm data.frame:
      write.csv(fpkm_DRG_NODOSE_filtered_df, file = "fpkm_DRG_NODOSE_filtered_list.txt")

      ### PCA analizi doğrudan RLD objesi kullanılarak yapılmıştı.

#      Daha sonra oluşturulan RLD matriksleri ise "NA" vs remove edildiği için,
#      PC1 ya PC2 genleri verilip heatmap oluşturulmak istendiğinde sıkıntı çıkardı.
#      Aşağıda heatmap öncesi o yüzden PC1 ve PC2 genleri içerisinde
#      "NA" olmayanlar önce belirlenip heatmap daha sonra oluşturuldu.

      ### Rld matrix for hetmaps, without non-number colums:

       rld_drg_nod_df_filtered_matrix = rld_drg_nod_df_filtered_matrix[,1:9]

      ###

      my_heatmap_labels$symbols = rld_drg_nod_df_filtered$symbols

       ### combine "symbol : gene name":

         temporary_x = paste(my_heatmap_labels$symbols, my_heatmap_labels$GENE_NAME, sep = " : ")
       head(temporary_x)


       my_heatmap_combined_names = as.data.frame(temporary_x)
       row.names(my_heatmap_combined_names) = row.names(my_heatmap_labels)

       head(my_heatmap_combined_names)

       colnames(my_heatmap_combined_names) = "gene_descriptions"
       head(my_heatmap_combined_names)
       rld_drg_nod_matrix_new_filtered = rld_drg_nod_matrix[row.names(my_heatmap_combined_names),]
       head(rld_drg_nod_matrix_new_filtered)

       head(rld_drg_nod_matrix_new_filtered[,1])
      length(rld_drg_nod_matrix_new_filtered[,1])
#      [1] 15864



    head(rld_drg_nod_df_filtered_matrix)


    ## Ayrıca matriksin elemanları numeric değil fakat sayıların hepsi karakter şeklinde.
    ## Bu yüzden matriks de tekrar oluşturulacak.

     ### Not: PCA analizinde kullanılan Top 500 genin hepsi son listede yok.
       ### O yüzden hata veriyor.

       tail(my_heatmap_combined_names[row.names(PC1_genes_ordered),] ,5)

    ## Ortak genlerin listesi:

     PC1_genes_intersect = intersect( row.names(PC1_genes_ordered), row.names(my_heatmap_combined_names))
     PC2_genes_intersect = intersect( row.names(PC2_genes_ordered), row.names(my_heatmap_combined_names))
     length(PC1_genes_intersect)
#    [1] 458
     length(PC2_genes_intersect)
#    [1] 458

    ### PC1 ve PC2 genes:

     heatmap.2(rld_drg_nod_matrix_new_filtered[PC1_genes_intersect[1:100],], scale="row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = my_heatmap_combined_names[PC1_genes_intersect[1:100],] ,margins = c(16, 21), dendrogram = c("column"), cexCol = 0.5, cexRow = 0.5)

     heatmap.2(rld_drg_nod_matrix_new_filtered[PC2_genes_intersect[1:100],], scale="row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = my_heatmap_combined_names[PC2_genes_intersect[1:100],] ,margins = c(16, 21), dendrogram = c("column"), cexCol = 0.5, cexRow = 0.5)


       heatmap.2(rld_drg_nod_matrix_new_filtered[PC2_genes_intersect[100:200],], scale="row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = my_heatmap_combined_names[PC2_genes_intersect[100:200],] ,margins = c(16, 21), dendrogram = c("column"), cexCol = 0.5, cexRow = 0.5)

       heatmap.2(rld_drg_nod_matrix_new_filtered[PC2_genes_intersect[200:300],], scale="row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = my_heatmap_combined_names[PC2_genes_intersect[200:300],] ,margins = c(16, 21), dendrogram = c("column"), cexCol = 0.5, cexRow = 0.5)

    ### EXPORTED as PDF US-legal format.

     heatmap.2(rld_drg_nod_matrix_new_filtered[PC2_genes_intersect[100:200],], scale="row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = my_heatmap_combined_names[PC2_genes_intersect[100:200],] ,margins = c(16, 21), dendrogram = c("column"), cexCol = 0.5, cexRow = 0.5)

       heatmap.2(rld_drg_nod_matrix_new_filtered[PC2_genes_intersect[200:300],], scale="row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = my_heatmap_combined_names[PC2_genes_intersect[200:300],] ,margins = c(16, 21), dendrogram = c("column"), cexCol = 0.5, cexRow = 0.5)


       heatmap.2(rld_drg_nod_matrix_new_filtered[PC2_genes_intersect[300:400],], scale="row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = my_heatmap_combined_names[PC2_genes_intersect[300:400],] ,margins = c(16, 21), dendrogram = c("column"), cexCol = 0.5, cexRow = 0.5)

       heatmap.2(rld_drg_nod_matrix_new_filtered[PC2_genes_intersect[400:458],], scale="row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = my_heatmap_combined_names[PC2_genes_intersect[400:458],] ,margins = c(16, 21), dendrogram = c("column"), cexCol = 0.5, cexRow = 0.5)

    ### END ####
    #######################################################################

#    Pseudogene ya da sembolü olmayan genleri ensembl İD ile tutmak için:

       rld_drg_nod_df_filtered$symbols = ifelse(is.na(rld_drg_nod_df_filtered$symbols), row.names(rld_drg_nod_df_filtered) , rld_drg_nod_df_filtered$symbols)

       rld_drg_nod_df_filtered$GENE_NAME = ifelse(is.na(rld_drg_nod_df_filtered$GENE_NAME), row.names(rld_drg_nod_df_filtered) , rld_drg_nod_df_filtered$GENE_NAME)

#    Bu komutlarda eğer sembol ya da gen ismi "NA" ise ensembl id'lerin yazılmasını
#sağlıyor.

#To check if "NA" symbols removed:

    setdiff(my_heatmap_labels$symbols, rld_drg_nod_df_filtered$symbols)
    #character(0)

    setdiff(my_heatmap_labels$GENE_NAME, rld_drg_nod_df_filtered$GENE_NAME)
    #character(0)

#Ancak daha önce tüm "NA" içeren sütunlara sahip satırları
#sildiğimiz için yaptuğımız değişiklikler listeleri etkilemedi.

    ### We will take original results and add symbols. But we will keep the rows without symbols.


     res_DRG_F_vs_cont_drg_symbols_NA_keeped = na.omit(res_DRG_F_vs_cont_drg) ###  all "NA"s are removed.

     res_NODS_F_vs_cont_nodose_symbols_NA_keeped = na.omit(res_NODS_F_vs_cont_nodose)

    ### Symbol addition:
     res_DRG_F_vs_cont_drg_symbols_NA_keeped$symbols = mapIds(org.Mm.eg.db, keys = row.names(res_DRG_F_vs_cont_drg_symbols_NA_keeped), keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first")
     res_NODS_F_vs_cont_nodose_symbols_NA_keeped$symbols = mapIds(org.Mm.eg.db, keys = row.names(res_NODS_F_vs_cont_nodose_symbols_NA_keeped), keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first")

    ### We will check if there is any "NA" symbols added.
    ### Then we will replace "NA" symbols with ensembl IDs.

     table(is.na(res_DRG_F_vs_cont_drg_symbols_NA_keeped$symbols))

#    FALSE  TRUE
#   15877  1990
    table(is.na(res_NODS_F_vs_cont_nodose_symbols_NA_keeped$symbols))

#    FALSE  TRUE
#    15877  1990

     ### ifelse(statement, (if TRUE: do this) ,(if FALSE: do that))

       res_DRG_F_vs_cont_drg_symbols_NA_keeped$symbols = ifelse(is.na(res_DRG_F_vs_cont_drg_symbols_NA_keeped$symbols), row.names(res_DRG_F_vs_cont_drg_symbols_NA_keeped), res_DRG_F_vs_cont_drg_symbols_NA_keeped$symbols)
     table(is.na(res_DRG_F_vs_cont_drg_symbols_NA_keeped$symbols))

    #FALSE
    #17867



     res_NODS_F_vs_cont_nodose_symbols_NA_keeped$symbols = ifelse(is.na(res_NODS_F_vs_cont_nodose_symbols_NA_keeped$symbols), row.names(res_NODS_F_vs_cont_nodose_symbols_NA_keeped), res_NODS_F_vs_cont_nodose_symbols_NA_keeped$symbols)
     table(is.na(res_NODS_F_vs_cont_nodose_symbols_NA_keeped$symbols))

    #FALSE
    #317867

    ### Additional Checks:

     anyNA(res_DRG_F_vs_cont_drg_symbols_NA_keeped, recursive=FALSE)
#    [1] FALSE
     anyNA(res_NODS_F_vs_cont_nodose_symbols_NA_keeped, recursive=FALSE)
#    [1] FALSE

     table(duplicated(res_DRG_F_vs_cont_drg_symbols_NA_keeped$symbols))

#    FALSE  TRUE
#    17843    24
     table(duplicated(res_NODS_F_vs_cont_nodose_symbols_NA_keeped$symbols))

#    FALSE  TRUE
#    17843    24

     #### REPLACE Step:

     ### ORDER by padj values:
     res_DRG_F_vs_cont_drg_symbols_NA_keeped = res_DRG_F_vs_cont_drg_symbols_NA_keeped[order( res_DRG_F_vs_cont_drg_symbols_NA_keeped$padj),]
     res_DRG_F_vs_cont_drg_symbols_NA_keeped
     res_NODS_F_vs_cont_nodose_symbols_NA_keeped = res_NODS_F_vs_cont_nodose_symbols_NA_keeped[order( res_NODS_F_vs_cont_nodose_symbols_NA_keeped$padj),]
     res_NODS_F_vs_cont_nodose_symbols_NA_keeped


     #### EXPORT files:
     write.csv(res_DRG_F_vs_cont_drg_symbols_NA_keeped, file = "res_DRG_F_vs_cont_drg_symbols_NA_keeped.txt")
     write.csv(res_NODS_F_vs_cont_nodose_symbols_NA_keeped, file = "res_NODS_F_vs_cont_nodose_symbols_NA_keeped.txt")

     ### NOT: Following results are with duplicated symbols:

      res_DRG_F_NA_keeped = subset(res_DRG_F_vs_cont_drg_symbols_NA_keeped, padj < 0.05)

      res_DRG_F_NA_keeped_sig = subset(res_DRG_F_NA_keeped, abs(log2FoldChange)  > 1.5)

      res_DRG_F_NA_keeped_sig_up = subset(res_DRG_F_NA_keeped_sig, log2FoldChange > 1.5)

      res_DRG_F_NA_keeped_sig_down = subset(res_DRG_F_NA_keeped_sig, log2FoldChange < -1.5)

     #####


      res_NODS_F_NA_keeped_sig = subset(res_NODS_F_vs_cont_nodose_symbols_NA_keeped, padj < 0.05)

      res_NODS_F_NA_keeped_sig = subset(res_NODS_F_NA_keeped_sig, abs(log2FoldChange) >  1.5)

      res_NODS_F_NA_keeped_sig_UP = subset(res_NODS_F_NA_keeped_sig, llog2FoldChange >  1.5)

      res_NODS_F_NA_keeped_sig_DOWN = subset(res_NODS_F_NA_keeped_sig, log2FoldChange < -1.5)

     ###


       table(duplicated(res_DRG_F_vs_cont_drg_symbols_NA_keeped$symbols))

#      FALSE  TRUE
#      17843    24
       table(duplicated(res_NODS_F_vs_cont_nodose_symbols_NA_keeped$symbols))

#      FALSE  TRUE
#      17843    24

      ## Duplike sembollerden en anlamlıları tutuldu. Ayrıca "NA" sembol olanlar
      ## ensembl id'leri ile tutuldu.

       res_DRG_F_vs_cont_drg_unique_symbols_NA_keeped = res_DRG_F_vs_cont_drg_symbols_NA_keeped[!duplicated(res_DRG_F_vs_cont_drg_symbols_NA_keeped$symbols),]

       res_NODS_F_vs_cont_nodose_unique_symbols_NA_keeped = res_NODS_F_vs_cont_nodose_symbols_NA_keeped[!duplicated(res_NODS_F_vs_cont_nodose_symbols_NA_keeped$symbols),]

      #### Heatmap için rld matrix vs. dosyalarını kontrol ettim:

       length(rld_drg_nod_matrix_new_filtered)
#      [1] 142776
       dim(rld_drg_nod_matrix_new_filtered)
#      [1] 15864     9
       length(rld_drg_nod_df_filtered$DRG_CS1_S4)
#      [1] 15864
       setdiff(row.names(rld_drg_nod_matrix_new_filtered), row.names(rld_drg_nod_df_filtered))
#      character(0)
       head(my_heatmap_combined_names)
       dim(my_heatmap_combined_names)
#      [1] 15864     1
       setdiff(row.names(rld_drg_nod_matrix_new_filtered), row.names(my_heatmap_combined_names))
#      character(0)

      ### rld matrix ve dataframe (df) nesnelerim doğru.

      ### Ayrıca heatmap label nesnesi de doğru.

      ### Heatmap için Nodoz ve DRG results nesnelerinde bulunan
      ### ortak ve farklı  genlere baktım:

       setdiff(row.names(res_DRG_F_vs_cont_drg_unique_symbols_NA_keeped), row.names(res_NODS_F_vs_cont_nodose_unique_symbols_NA_keeped))

      ## Bu 11 gen DRG listesinde var sadece.

       setdiff(row.names(res_NODS_F_vs_cont_nodose_unique_symbols_NA_keeped),row.names(res_DRG_F_vs_cont_drg_unique_symbols_NA_keeped))

      ## Bu 11 gen sadece nodoz listesinde var sadece.

      ### Bu genlerin hangi gen sembollerine karşılık geldiğine bakmak için:

       res_NODS_F_vs_cont_nodose_unique_symbols_NA_keeped[setdiff(row.names(res_NODS_F_vs_cont_nodose_unique_symbols_NA_keeped),row.names(res_DRG_F_vs_cont_drg_unique_symbols_NA_keeped)),]
       res_DRG_F_vs_cont_drg_unique_symbols_NA_keeped[setdiff(row.names(res_DRG_F_vs_cont_drg_unique_symbols_NA_keeped), row.names(res_NODS_F_vs_cont_nodose_unique_symbols_NA_keeped)),]

       ### Same genes but different ensembl ids in DRG and Nodose.
         ### Only Hbb-b1 gene is significant.

         ### We will form a list, which includes genes found at both nodose and drg lists:

         DRG_nodose_intersection_genes = intersect(row.names(res_DRG_F_vs_cont_drg_unique_symbols_NA_keeped), row.names(res_NODS_F_vs_cont_nodose_unique_symbols_NA_keeped))
         length(DRG_nodose_intersection_genes)
#        [1] 17832

      ### rld prepare for heatmap:

       head(rld_drg_nod_matrix)
       rld_drg_nod_matrix_last = rld_drg_nod_matrix[DRG_nodose_intersection_genes,]
       head(rld_drg_nod_matrix_last)
       dim(rld_drg_nod_matrix_last)
#      [1] 17832     9

      ### ready for topVAR.

       #### For row labels of heatmap:
         #### "my_heatmap_combined_names["ensmebl ids of select genes", ]"
         ### will be used.



        dim(fpkm_DRG_NODOSE_run_1)
#      [1] 25202    11
        length(intersect(row.names(res_NODS_F_vs_cont_nodose_unique_symbols_NA_keeped), row.names(fpkm_DRG_NODOSE_run_1)))
#      [1] 17843
        length(intersect(row.names(res_DRG_F_vs_cont_drg_unique_symbols_NA_keeped), row.names(fpkm_DRG_NODOSE_run_1)))
#       [1] 17843
        ### DRG'nin 17843 geni baz alınacak.
          ### Bu matrix Hbb-b1 içerenler için.
         ### DRG baz alınarak hazırlanan FPKM matriksi
         ### genel için de kullanılacak.
         ### Ayrıca TopVAR için kullanılacak.
          ### Sadece nodoz için FPKM hazırlanacak.
         ### Bu nodozda ekspres edilen genlere bakmak için olacak.


         #####################Prepare matrixes for later TopVAR heatmap:

          dim(rld_drg_nod_matrix)
#       [1] 25202     9
        rld_DRG_results_filtered = rld_drg_nod_matrix[row.names(res_DRG_F_vs_cont_drg_unique_symbols_NA_keeped),]

          ####
          ### For nodose:
          rld_Nodose_results_filtered = rld_drg_nod_matrix[row.names(res_NODS_F_vs_cont_nodose_unique_symbols_NA_keeped),]

          ###
          ### we formed an fpkm list which contains ensembl ids in DRG results list.
          ### secondly, we did same for nodose.
          ### It is checked and DRG fpkm list can be used for TopVAR analysis.
         ####


         ### FPKM prepare:

         fpkm_drg_nod_matrix_w_padj = fpkm_DRG_NODOSE_run_1[DRG_nodose_intersection_genes,]

       ### by following command, we order drg results with fpkm list-row name order:

        drg_template = res_DRG_F_vs_cont_drg_unique_symbols_NA_keeped[DRG_nodose_intersection_genes,]
        head(drg_template)


        nodose_template = res_NODS_F_vs_cont_nodose_unique_symbols_NA_keeped[DRG_nodose_intersection_genes,]
        head(nodose_template)


       ### Adding padj values to the fpkm list:

        fpkm_drg_nod_matrix_w_padj$DRG_Padj = drg_template$padj
        class(fpkm_drg_nod_matrix_w_padj)
#       [1] "data.frame"
        head(fpkm_drg_nod_matrix_w_padj)


        fpkm_drg_nod_matrix_w_padj$NODOSE_Padj = nodose_template$padj
        head(fpkm_drg_nod_matrix_w_padj)

        write.csv(fpkm_drg_nod_matrix_w_padj, file = "fpkm_drg_nod_matrix_w_padj.txt")

        ### NOTLAR:

#       Drg ve nodoz ortak genler, intersect fonksiyonu kullanılarak ve
#       drg gen isimleri ilk input verilerek hazırlandı.

#       Bu yüzden "DRG_nodose_intersection_genes" vektörü genleri DRG sırasına
#       göre içermektedir. Bu beklenen bir durum.

####

# Yukarıda drg ve nodoz results listelerindeki bilgiler, "DRG_nodose_intersection_genes" vektöründeki
# genler için template değişkenlere eşitlendi. Bu listelerin sıralaması fpkm listesi
# ile aynı olduğu için, bu template dosyalarının p-adjusted değerleri fpkm data-frame listesine
# sütun olarak eklenebildi.

### END ###

### sadece DRG ve sadece nodoz results listesinde olan genler için ayrı fpkm hazırlama:
### daha önceki fpkm yöntemi ile benzer adımlar uygulandı.
### DRG ve Nodoz için unique olan 11 genin ismi alınıp, sonra bunlara ait fpkm listesi hazırlandı.
### Daha sonra aynı isimlerin padj değerleri alınıp fpkm listelerine eklendi.
### Bunlar daha sonra exce üzerinde ana fpkm listesine eklenecek.

 ### START:
 drg_only_genes = setdiff(row.names(res_DRG_F_vs_cont_drg_unique_symbols_NA_keeped), row.names(res_NODS_F_vs_cont_nodose_unique_symbols_NA_keeped))
 nodose_only_genes = setdiff(row.names(res_NODS_F_vs_cont_nodose_unique_symbols_NA_keeped),row.names(res_DRG_F_vs_cont_drg_unique_symbols_NA_keeped))
 ###
  dim(fpkm_DRG_NODOSE_run_1)
#       [1] 25202    11
 ### this is main fpkm list
  head(fpkm_DRG_NODOSE_run_1)



 fpkm_drg_only_11_genes = fpkm_DRG_NODOSE_run_1[drg_only_genes,]
 fpkm_nodose_only_11_genes = fpkm_DRG_NODOSE_run_1[nodose_only_genes,]
 ###
   drg_template_2 = res_DRG_F_vs_cont_drg_unique_symbols_NA_keeped[drg_only_genes,]
 nodose_template_2 = res_NODS_F_vs_cont_nodose_unique_symbols_NA_keeped[nodose_only_genes,]
 ####
   fpkm_drg_only_11_genes$drg_padj = drg_template_2$padj
 fpkm_nodose_only_11_genes$nodose_padj = nodose_template_2$padj
 ####
   fpkm_drg_only_11_genes
fpkm_nodose_only_11_genes


 ### all is checked and confirmed.
   write.csv(fpkm_drg_only_11_genes, file = "fpkm_drg_only_11_genes.txt")
 write.csv(fpkm_nodose_only_11_genes, file = "fpkm_nodose_only_11_genes.txt")
 ###


 ##############################################################################################################################

 ### Visualisation of DESEQ2 Dataset with vidger package.


  library(vidger)
  library(DESeq2)

 ### Example dataset:

   df.deseq

 # class: DESeqDataSet
 # dim: 9556 6
 # metadata(1): version
 # assays(3): counts mu cooks
 # rownames(9556): FBgn0000008 FBgn0000017 ... FBgn0261574 FBgn0261575
 # rowData names(25): baseMean baseVar ... deviance maxCooks
 # colnames(6): treated2 treated3 ... untreated3 untreated4
 # colData names(2): condition sizeFactor

 ### Our dataset:

 dds_drg_nod

 # class: DESeqDataSet
 # dim: 25202 9
 # metadata(1): version
 # assays(5): counts avgTxLength normalizationFactors mu cooks
 # rownames(25202): ENSMUSG00000000001 ENSMUSG00000000028 ... ENSMUSG00000115847 ENSMUSG00000115848
 # rowData names(29): baseMean baseVar ... deviance maxCooks
 # colnames(9): DRG_CS1_S4 DRG_CS2_S5 ... NG_F_S6 NG_F_S7
 # colData names(2): SampleNames condition

 ### PLOTS:

 ### For DE genes:

  vsDEGMatrix(data =  dds_drg_nod, padj = 0.05, d.factor = 'condition',type = 'deseq', title = TRUE, legend = TRUE, grid = TRUE)

  vsMAPlot(x = 'cont_drg', y = 'sample_drg',data = dds_drg_nod, d.factor = 'condition', type = 'deseq',padj = 0.05, y.lim = NULL, lfc = 1.5, title = TRUE,legend = TRUE, grid = TRUE)

  vsMAPlot(x = 'cont_nodose', y = 'sample_nodose',data = dds_drg_nod, d.factor = 'condition', type = 'deseq',padj = 0.05, y.lim = NULL, lfc = 1.5, title = TRUE,legend = TRUE, grid = TRUE)

  vsVolcano(x = 'cont_drg', y = 'sample_drg',data = dds_drg_nod, d.factor = 'condition', type = 'deseq', padj = 0.05, x.lim = NULL, lfc = 1.5, title = TRUE, legend = TRUE, grid = TRUE, data.return = FALSE)

  vsVolcano(x = 'cont_nodose', y = 'sample_nodose',data = dds_drg_nod, d.factor = 'condition', type = 'deseq', padj = 0.05, x.lim = NULL, lfc = 1.5, title = TRUE, legend = TRUE, grid = TRUE, data.return = FALSE)


  #####################################################################################################

 ####
 #### PCA with new names:


  library(pca3d)

  pca3d(pca, components = 1:3, axes.color = "grey", bg = "white", radius = 1, show.labels = c("DRG-T1", "DRG-T2", "DRG-HS2", "DRG-HS1", "DRG-HS3","NG-T1", "NG-T2", "NG-HS1", "NG-HS-2"), axe.titles = c(paste0("PC1: %", percentVar_pca3d[1]),paste0("PC2: %", percentVar_pca3d[2]),paste0("PC3: %", percentVar_pca3d[3])) ,show.axe.titles = TRUE, col = c("blue","blue", "yellow","yellow", "yellow", "green", "green", "red", "red"))
#        [1] 0.6663044 0.3395323 0.1732928
#        Creating new device

 ### bigger radius of balls:

  pca3d(pca, components = 1:3, axes.color = "grey", bg = "white", radius = 2, show.labels = c("DRG-T1", "DRG-T2", "DRG-HS2", "DRG-HS1", "DRG-HS3","NG-T1", "NG-T2", "NG-HS1", "NG-HS-2"), axe.titles = c(paste0("PC1: %", percentVar_pca3d[1]),paste0("PC2: %", percentVar_pca3d[2]),paste0("PC3: %", percentVar_pca3d[3])) ,show.axe.titles = TRUE, col = c("blue","blue", "yellow","yellow", "yellow", "green", "green", "red", "red"))
#         [1] 1.3326088 0.6790645 0.3465857
#         Creating new device


  ################################################################################################

  # Heatmap visualisation of top variant genes.

  head(rld_DRG_results_filtered)


   dim(rld_DRG_results_filtered)
#          [1] 17843     9
   head(my_heatmap_combined_names)

  ### TopVAR
   topVarGenes <- head(order(rowVars(rld_DRG_results_filtered),decreasing=TRUE),100)
   mat <- rld_DRG_results_filtered[topVarGenes,]
   mat <- mat - rowMeans(mat)
   library(gplots)

   heatmap.2(mat, scale="row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = my_heatmap_combined_names[row.names(mat),] ,margins = c(16, 21), dendrogram = 'col' , cexRow = 0.4, cexCol = 0.75)

  ### plot exported as us legal pdf with 8 x 14 inches.

  #### NOT: My default heatmap parameters are as below:

   heatmap.2(mat, scale="row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = gene_labels[topVarGenes] ,margins = c(16, 21), dendrogram = c("column"), cexCol = 0.5, cexRow = 0.5)

  ### cexrow and cexcol are used to specify label fonts.

  ### plot exported as us legal pdf with 8 x 16 inches.



#################################################################################################################


   #### Retrieval of molecular signature database for mouse.

   ### we will get the msigdb gene sets.

   ### GO TERMS for mouse in a dataframe with entrez ids:

    library(msigdf)
    dim(msigdf.mouse)
#   [1] 3912066       3
    head(msigdf.mouse,5)
   # A tibble: 5 x 3
#   collection geneset    entrez
#   <chr      <chr       <int
   # 1 c1         chr5q23 100042150
   # 2 c1         chr5q23    103236
   # 3 c1         chr5q23    106869
   # 4 c1         chr5q23    107022
   # 5 c1         chr5q23    109700

    library(tidyverse)

    msigdf.mouse %% head()
   # A tibble: 6 x 3
   # collection geneset    entrez
   # <chr      <chr       <int
   #   1 c1         chr5q23 100042150
   # 2 c1         chr5q23    103236
   # 3 c1         chr5q23    106869
   # 4 c1         chr5q23    107022
   # 5 c1         chr5q23    109700
   # 6 c1         chr5q23     11548

    mouse_c5_v5.2_df = (msigdf.mouse %% filter(collection=="c5"))
    mouse_c5_v5.2_df = as.data.frame(mouse_c5_v5.2_df)
    head( mouse_c5_v5.2_df)
   #
   # collection                                     geneset entrez
   # 1         c5 GO_REGULATION_OF_DOPAMINE_METABOLIC_PROCESS 108015
   # 2         c5 GO_REGULATION_OF_DOPAMINE_METABOLIC_PROCESS 109731
   # 3         c5 GO_REGULATION_OF_DOPAMINE_METABOLIC_PROCESS  11444
   # 4         c5 GO_REGULATION_OF_DOPAMINE_METABOLIC_PROCESS  12846
   # 5         c5 GO_REGULATION_OF_DOPAMINE_METABOLIC_PROCESS  13162
   # 6         c5 GO_REGULATION_OF_DOPAMINE_METABOLIC_PROCESS  13488


    dim(mouse_c5_v5.2_df)
#   [1] 1013231       3


    ### our MsigDF is ready to be used!!!!

####################################################################################################################

  #### FPKM MATRICES FOR GENES WITH SPECIFIC FUNCTIONS.

    ### NOT: DESEQ2 tarafından filtrelenmiş results listesindeki
    ### genleri içeren fpkm listesi kullanıldı.

     dim(fpkm_drg_nod_matrix_w_padj)
#    [1] 17832    13

    ### To subset data frame for spesific gene symbols or gene names:

    selected_TRP_genes = fpkm_drg_nod_matrix_w_padj[grepl("transient receptor potential", fpkm_drg_nod_matrix_w_padj$gene_name , ignore.case = TRUE),]

    selected_TRP_genes

    selected_purinergic_receptor_genes = fpkm_drg_nod_matrix_w_padj[grepl("purinergic receptor", fpkm_drg_nod_matrix_w_padj$gene_name , ignore.case = TRUE),]

    selected_TRP_genes

    selected_purinergic_receptor_genes

    selected_mechano_genes = fpkm_drg_nod_matrix_w_padj[grepl("mechano", fpkm_drg_nod_matrix_w_padj$gene_name , ignore.case = TRUE),]

    selected_mechano_genes

    selected_channel_genes = fpkm_drg_nod_matrix_w_padj[grepl("channel", fpkm_drg_nod_matrix_w_padj$gene_name , ignore.case = TRUE),]

    selected_channel_genes

    selected_gpr_genes = fpkm_drg_nod_matrix_w_padj[grepl("GPR", fpkm_drg_nod_matrix_w_padj$gene_name , ignore.case = TRUE),]

    selected_G_protein_coupled_genes = fpkm_drg_nod_matrix_w_padj[grepl("G protein-coupled", fpkm_drg_nod_matrix_w_padj$gene_name , ignore.case = FALSE),]

#############################################################################################################################


    #### Producing heatmaps for DE genes represented in spesifc GO TERMS:



  library(goseq)


  all_GO_BP_terms_for_DRG_NODOSE <- stack(getgo(row.names(fpkm_DRG_NODOSE_run_1), 'mm10', 'ensGene', fetch.cats = "GO:BP"))



  all_GO_MF_terms_for_DRG_NODOSE <- stack(getgo(row.names(fpkm_DRG_NODOSE_run_1), 'mm10', 'ensGene', fetch.cats = "GO:MF"))

  ### Obtain GO ID-GO TERM df:

  library(GO.db)
  go_terms = Term(GOTERM)
  head(go_terms)


  ### Finding GO.term of interest:

    go_terms_df["GO:0004930",]
  #  [1] G-protein coupled receptor activity
  #  44917 Levels: 'de novo' actin filament nucleation ...

  go_terms_df %% head()

  # go_terms
  # GO:0000001                                mitochondrion inheritance
  # GO:0000002                         mitochondrial genome maintenance
  # GO:0000003                                             reproduction
  # GO:0000006    high-affinity zinc transmembrane transporter activity
  # GO:0000007 low-affinity zinc ion transmembrane transporter activity
  # GO:0000009                   alpha-1,6-mannosyltransferase activity
  #

  ########################################################################################################
  ### get names of GO ids:

   go_terms_df["GO:0003700",]
  # [1] DNA binding transcription factor activity
  # 44917 Levels: 'de novo' actin filament nucleation ...

  ### take all gene ids in specific GO terms:

   DNA_binding_TF_activity_GO_0003700 =  subset(all_GO_MF_terms_for_DRG_NODOSE, all_GO_MF_terms_for_DRG_NODOSE$values == "GO:0003700")
   dim(DNA_binding_TF_activity_GO_0003700)
  # [1] 1179    2

   GPCR_activity_GO_0004930 =  subset(all_GO_MF_terms_for_DRG_NODOSE, all_GO_MF_terms_for_DRG_NODOSE$values == "GO:0004930")
   dim(GPCR_activity_GO_0004930)
  # [1] 507   2

   ion_channel_activity_GO_0005216 =  subset(all_GO_MF_terms_for_DRG_NODOSE, all_GO_MF_terms_for_DRG_NODOSE$values == "GO:0005216")
   dim(ion_channel_activity_GO_0005216)
  # [1] 392   2

   ####
     GPCR_activity_genes = as.character(GPCR_activity_GO_0004930$ind)

  DNA_binding_TF_activity_genes = as.character(DNA_binding_TF_activity_GO_0003700$ind)

  ion_channel_activity_genes = as.character(ion_channel_activity_GO_0005216$ind)

  ### Prepare significant lists:

  res_NODS_F_vs_cont_nodose_symbols_NA_keeped_sig = subset(res_NODS_F_vs_cont_nodose_symbols_NA_keeped, padj < 0.05)
  res_NODS_F_vs_cont_nodose_symbols_NA_keeped_sig_log05_UP = subset(res_NODS_F_vs_cont_nodose_symbols_NA_keeped_sig, log2FoldChange > 0.5)
  res_NODS_F_vs_cont_nodose_symbols_NA_keeped_sig_log05_DOWN = subset(res_NODS_F_vs_cont_nodose_symbols_NA_keeped_sig, log2FoldChange < -0.5)

  res_DRG_F_vs_cont_drg_symbols_NA_keeped_sig = subset(res_DRG_F_vs_cont_drg_symbols_NA_keeped, padj < 0.05)
  res_DRG_F_vs_cont_drg_symbols_NA_keeped_sig_log05_UP = subset(res_DRG_F_vs_cont_drg_symbols_NA_keeped_sig, log2FoldChange >  0.5)
  res_DRG_F_vs_cont_drg_symbols_NA_keeped_sig_log05_DOWN = subset(res_DRG_F_vs_cont_drg_symbols_NA_keeped_sig, log2FoldChange < -0.5)

  ### Find genes associated with GO terms:

   DRG_05_UP_GPCR_genes =intersect(row.names(res_DRG_F_vs_cont_drg_symbols_NA_keeped_sig_log05_UP), GPCR_activity_genes)
   DRG_05_DOWN_GPCR_genes =intersect(row.names(res_DRG_F_vs_cont_drg_symbols_NA_keeped_sig_log05_DOWN), GPCR_activity_genes)
   ###
     DRG_05_UP_ion_channel_genes =intersect(row.names(res_DRG_F_vs_cont_drg_symbols_NA_keeped_sig_log05_UP), ion_channel_activity_genes)
   DRG_05_DOWN_ion_channel_genes =intersect(row.names(res_DRG_F_vs_cont_drg_symbols_NA_keeped_sig_log05_DOWN), ion_channel_activity_genes)
   ###
     DRG_05_UP_TFactor_genes =intersect(row.names(res_DRG_F_vs_cont_drg_symbols_NA_keeped_sig_log05_UP), DNA_binding_TF_activity_genes)
   DRG_05_DOWN_TFactor_genes =intersect(row.names(res_DRG_F_vs_cont_drg_symbols_NA_keeped_sig_log05_DOWN), DNA_binding_TF_activity_genes)
   ###
   Nodose_05_UP_GPCR_genes =intersect(row.names(res_NODS_F_vs_cont_nodose_symbols_NA_keeped_sig_log05_UP), GPCR_activity_genes)
   Nodose_05_DOWN_GPCR_genes =intersect(row.names(res_NODS_F_vs_cont_nodose_symbols_NA_keeped_sig_log05_DOWN), GPCR_activity_genes)
   ###
     Nodose_05_UP_ion_channel_genes =intersect(row.names(res_NODS_F_vs_cont_nodose_symbols_NA_keeped_sig_log05_UP), ion_channel_activity_genes)
   Nodose_05_DOWN_ion_channel_genes =intersect(row.names(res_NODS_F_vs_cont_nodose_symbols_NA_keeped_sig_log05_DOWN), ion_channel_activity_genes)
   ###
     Nodose_05_UP_TFactor_genes =intersect(row.names(res_NODS_F_vs_cont_nodose_symbols_NA_keeped_sig_log05_UP), DNA_binding_TF_activity_genes)
   Nodose_05_DOWN_TFactor_genes =intersect(row.names(res_NODS_F_vs_cont_nodose_symbols_NA_keeped_sig_log05_DOWN), DNA_binding_TF_activity_genes)
   ####

     ####
     length(DRG_05_UP_GPCR_genes)
#  [1] 5
   length(DRG_05_DOWN_GPCR_genes)
#  [1] 20
   length(DRG_05_UP_ion_channel_genes)
#  [1] 7
   length(DRG_05_DOWN_ion_channel_genes)
#  [1] 10
   length(DRG_05_UP_TFactor_genes)
#  [1] 26
   length(DRG_05_DOWN_TFactor_genes)
#  [1] 80

   ### all significant genes together:
     DRG_05_sig_all_GPCR_genes =intersect(row.names(res_DRG_F_vs_cont_drg_symbols_NA_keeped_sig_log05_ALL), GPCR_activity_genes)
   length(DRG_05_sig_all_GPCR_genes)
#  [1] 25
   DRG_05_sig_all_ion_channel_genes =intersect(row.names(res_DRG_F_vs_cont_drg_symbols_NA_keeped_sig_log05_ALL), ion_channel_activity_genes)
   DRG_05_sig_all_TF_genes =intersect(row.names(res_DRG_F_vs_cont_drg_symbols_NA_keeped_sig_log05_ALL), DNA_binding_TF_activity_genes)
   ###
     Nodose_05_sig_all_GPCR_genes =intersect(row.names(res_NODS_F_vs_cont_nodose_symbols_NA_keeped_sig_log05_ALL), GPCR_activity_genes)
   Nodose_05_sig_all_ion_channel_genes =intersect(row.names(res_NODS_F_vs_cont_nodose_symbols_NA_keeped_sig_log05_ALL), ion_channel_activity_genes)
   Nodose_05_sig_all_TF_genes =intersect(row.names(res_NODS_F_vs_cont_nodose_symbols_NA_keeped_sig_log05_ALL), DNA_binding_TF_activity_genes)
   ###

    class(rld_DRG_results_filtered)
#  [1] "matrix"
   head(rld_DRG_results_filtered)

   head(rld_DRG_results_filtered[,1:5])



   head(genes_symbols_df)
  genes_symbols_df
  # ENSMUSG00000000001            Gnai3
  # ENSMUSG00000000028            Cdc45
  # ENSMUSG00000000037            Scml2
  # ENSMUSG00000000049             Apoh
  # ENSMUSG00000000056             Narf
  # ENSMUSG00000000058             Cav2
   dim(genes_symbols_df)
  # [1] 25202     1


  ### combine two vector:

   c(DRG_05_UP_GPCR_genes,DRG_05_DOWN_GPCR_genes)


  # [1] "ENSMUSG00000033470" "ENSMUSG00000078698" "ENSMUSG00000070552"
  # [4] "ENSMUSG00000028563" "ENSMUSG00000070550" "ENSMUSG00000039167"
  # [7] "ENSMUSG00000059588" "ENSMUSG00000021319" "ENSMUSG00000045730"
  # [10] "ENSMUSG00000044338" "ENSMUSG00000040125" "ENSMUSG00000022297"
  # [13] "ENSMUSG00000049791" "ENSMUSG00000045092" "ENSMUSG00000044337"
  # [16] "ENSMUSG00000033446" "ENSMUSG00000020123" "ENSMUSG00000003974"
  # [19] "ENSMUSG00000043953" "ENSMUSG00000061878" "ENSMUSG00000028782"
  # [22] "ENSMUSG00000001240" "ENSMUSG00000056492" "ENSMUSG00000050511"
  # [25] "ENSMUSG00000021070"

  ###
   ### heatmaps:


     library(gplots)

   heatmap.2(rld_DRG_results_filtered[,1:5][c(DRG_05_UP_GPCR_genes,DRG_05_DOWN_GPCR_genes),], scale = "row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = genes_symbols_df[c(DRG_05_UP_GPCR_genes,DRG_05_DOWN_GPCR_genes),] ,margins = c(16, 21) ,dendrogram = "none", Rowv = NULL ,cexCol = 0.5, cexRow = 0.5)


   heatmap.2(rld_DRG_results_filtered[,1:5][c(DRG_05_UP_ion_channel_genes,DRG_05_DOWN_ion_channel_genes),], scale = "row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = genes_symbols_df[c(DRG_05_UP_ion_channel_genes,DRG_05_DOWN_ion_channel_genes),] ,margins = c(16, 21) ,dendrogram = "none", Rowv = NULL ,cexCol = 0.5, cexRow = 0.5)
   ###
     heatmap.2(rld_DRG_results_filtered[,1:5][c(DRG_05_UP_TFactor_genes,DRG_05_DOWN_TFactor_genes),], scale = "row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = genes_symbols_df[c(DRG_05_UP_TFactor_genes,DRG_05_DOWN_TFactor_genes),] ,margins = c(16, 21) ,dendrogram = "none", Rowv = NULL ,cexCol = 0.5, cexRow = 0.5)

  ### Without key box:

   heatmap.2(rld_DRG_results_filtered[,1:5][c(DRG_05_UP_TFactor_genes,DRG_05_DOWN_TFactor_genes),], scale = "row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = genes_symbols_df[c(DRG_05_UP_TFactor_genes,DRG_05_DOWN_TFactor_genes),] ,margins = c(16, 21) ,dendrogram = "none", Rowv = NULL ,cexCol = 0.5, cexRow = 0.5, key = FALSE)

   ###
     ### not: drg_results_filtered_rld matriksi nodoz için de uygundur.
     ###

     ### colw = NULL kullanılacak çünkü nodoz-F sağda olacak.
     heatmap.2(rld_DRG_results_filtered[,6:9][c(Nodose_05_UP_GPCR_genes,Nodose_05_DOWN_GPCR_genes),], scale = "row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = genes_symbols_df[c(Nodose_05_UP_GPCR_genes,Nodose_05_DOWN_GPCR_genes),] ,margins = c(16, 21) ,dendrogram = "none", Rowv = NULL, Colv = NULL ,cexCol = 0.5, cexRow = 0.5)
   ###
     heatmap.2(rld_DRG_results_filtered[,6:9][c(Nodose_05_UP_ion_channel_genes,Nodose_05_DOWN_ion_channel_genes),], scale = "row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = genes_symbols_df[c(Nodose_05_UP_ion_channel_genes,Nodose_05_DOWN_ion_channel_genes),] ,margins = c(16, 21) ,dendrogram = "none", Rowv = NULL, Colv = NULL ,cexCol = 0.5, cexRow = 0.5)
   heatmap.2(rld_DRG_results_filtered[,6:9][c(Nodose_05_UP_TFactor_genes,Nodose_05_DOWN_TFactor_genes),], scale = "row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = genes_symbols_df[c(Nodose_05_UP_TFactor_genes,Nodose_05_DOWN_TFactor_genes),] ,margins = c(16, 21) ,dendrogram = "none", Rowv = NULL, Colv = NULL ,cexCol = 0.5, cexRow = 0.5)


##################################################################################################


   #### HETAMAPS

   ######### NOT: "ColV = NULL" komutuna gerek kalmadı. Fakat bundan sonra default olarak eklenecek, sıralı heatmap için.
    heatmap.2(rld_DRG_results_filtered[,1:5][c(DRG_05_UP_TFactor_genes,DRG_05_DOWN_TFactor_genes),], scale = "row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = genes_symbols_df[c(DRG_05_UP_TFactor_genes,DRG_05_DOWN_TFactor_genes),] ,margins = c(1, 16) ,dendrogram = "none", Rowv = NULL ,cexCol = 0.5, cexRow = 1, key = FALSE, ColSideColors = as.character(c("blue", "blue", "yellow", "yellow","yellow")), labCol = FALSE)
   ### (Device size) 8.5 x 16 exported.
   ### tiff: 722x1500


     ####

     heatmap.2(rld_DRG_results_filtered[,1:5][c(DRG_05_UP_GPCR_genes,DRG_05_DOWN_GPCR_genes),], scale = "row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = genes_symbols_df[c(DRG_05_UP_GPCR_genes,DRG_05_DOWN_GPCR_genes),] ,margins = c(1, 16) ,dendrogram = "none", Rowv = NULL ,cexCol = 0.5, cexRow = 1, key = FALSE, ColSideColors = as.character(c("blue", "blue", "yellow", "yellow","yellow")), labCol = FALSE)

    ### export tiff: 722x484
     ### pdf: (Device size)- 7.52x5.89


    heatmap.2(rld_DRG_results_filtered[,1:5][c(DRG_05_UP_ion_channel_genes,DRG_05_DOWN_ion_channel_genes),], scale = "row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = genes_symbols_df[c(DRG_05_UP_ion_channel_genes,DRG_05_DOWN_ion_channel_genes),] ,margins = c(1, 16) ,dendrogram = "none", Rowv = NULL ,cexCol = 0.5, cexRow = 1, key = FALSE, ColSideColors = as.character(c("blue", "blue", "yellow", "yellow","yellow")), labCol = FALSE)

     ### export tiff: 722x484
     ### pdf: (Device size)- 7.52x5.89

     ### Legend:

     plot.new()

   par(lend = 1)
   legend("topright", legend = c("DRG-T1", "DRG-T2", "DRG-HS2", "DRG-HS1", "DRG-HS3") , col = c("blue", "blue", "yellow", "yellow","yellow"),lty= 1,lwd = 10 )

   ### export tiff: 722x600

   #############################################################################
   #############################################################################

   ### NODOSE heatmaps. NOT: "ColV = NULL" command is added.

   heatmap.2(rld_DRG_results_filtered[,6:9][c(Nodose_05_UP_ion_channel_genes,Nodose_05_DOWN_ion_channel_genes),], scale = "row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = genes_symbols_df[c(Nodose_05_UP_ion_channel_genes,Nodose_05_DOWN_ion_channel_genes),] ,margins = c(1, 16) ,dendrogram = "none", Rowv = NULL ,cexCol = 0.5, cexRow = 1, key = FALSE, ColSideColors = as.character(c("blue", "blue", "yellow", "yellow")), labCol = FALSE, Colv = NULL)


   ### export tiff: 722x484
    ### pdf: (Device size)- 7.52x5.89


     heatmap.2(rld_DRG_results_filtered[,6:9][c(Nodose_05_UP_GPCR_genes,Nodose_05_DOWN_GPCR_genes),], scale = "row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = genes_symbols_df[c(Nodose_05_UP_GPCR_genes,Nodose_05_DOWN_GPCR_genes),] ,margins = c(1, 16) ,dendrogram = "none", Rowv = NULL ,cexCol = 0.5, cexRow = 1, key = FALSE, ColSideColors = as.character(c("blue", "blue", "yellow", "yellow")), labCol = FALSE, Colv = NULL)

   ### export tiff: 722x484
     ### pdf: (Device size)- 7.52x5.89
     ###
     heatmap.2(rld_DRG_results_filtered[,6:9][c(Nodose_05_UP_TFactor_genes,Nodose_05_DOWN_TFactor_genes),], scale = "row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = genes_symbols_df[c(Nodose_05_UP_TFactor_genes,Nodose_05_DOWN_TFactor_genes),] ,margins = c(1, 16) ,dendrogram = "none", Rowv = NULL ,cexCol = 0.5, cexRow = 1, key = FALSE, ColSideColors = as.character(c("blue", "blue", "yellow", "yellow")), labCol = FALSE,  Colv = NULL)


     ### (Device size)-  8.5 x 16 exported.
     ### tiff: 722x1500

      ### labels:
  plot.new()
  par(lend = 1)
  legend("topright", legend = c("NG-T1", "NG-T2", "NG-HS1", "NG-HS-2") , col = c(,"blue", "blue", "yellow", "yellow"),lty= 1,lwd = 10 )


  ### export tiff: 722x600


  ### Prepare Top 50 genes:

   DRG_F_sig_Log05_up_genes = head(row.names(res_DRG_F_vs_cont_drg_symbols_NA_keeped_sig_log05_UP),50)
   DRG_F_sig_Log05_down_genes = head(row.names(res_DRG_F_vs_cont_drg_symbols_NA_keeped_sig_log05_DOWN),50)
   ####
     Nodose_F_sig_Log05_up_genes = head(row.names(res_NODS_F_vs_cont_nodose_symbols_NA_keeped_sig_log05_UP),50)
   Nodose_F_sig_Log05_down_genes = head(row.names(res_NODS_F_vs_cont_nodose_symbols_NA_keeped_sig_log05_DOWN),50)
   ###########

  ### rename matrix with more reasonable one:

   rld_filtered_last_matrix = rld_DRG_results_filtered

  #### Versions with "NA" symbols replaced with ENS ids:

   head(genes_symbols_df)
  genes_symbols_df
  # ENSMUSG00000000001            Gnai3
  # ENSMUSG00000000028            Cdc45
  # ENSMUSG00000000037            Scml2
  # ENSMUSG00000000049             Apoh
  # ENSMUSG00000000056             Narf
  # ENSMUSG00000000058             Cav2
  #
   table(is.na(genes_symbols_df$genes_symbols_df))

  # FALSE  TRUE
  # 18563  6639
  #
   genes_symbols_NA_symbols_replaced_w_ENS_ids = genes_symbols_df

  ##

     colnames(genes_symbols_NA_symbols_replaced_w_ENS_ids)
  # [1] "genes_symbols_df"
  #
   class(genes_symbols_NA_symbols_replaced_w_ENS_ids$genes_symbols_df)
  # [1] "factor"

  colnames(genes_symbols_NA_symbols_replaced_w_ENS_ids) = c("symbols")

   head(genes_symbols_NA_symbols_replaced_w_ENS_ids)
  symbols
  # ENSMUSG00000000001   Gnai3
  # ENSMUSG00000000028   Cdc45
  # ENSMUSG00000000037   Scml2
  # ENSMUSG00000000049    Apoh
  # ENSMUSG00000000056    Narf
  # ENSMUSG00000000058    Cav2

  ### symbols should be characters, not factors, to apply ifelse function:

   genes_symbols_NA_symbols_replaced_w_ENS_ids$symbols = as.character(genes_symbols_NA_symbols_replaced_w_ENS_ids$symbols)
   ###
     genes_symbols_NA_symbols_replaced_w_ENS_ids$symbols = ifelse(is.na(genes_symbols_NA_symbols_replaced_w_ENS_ids$symbols), row.names(genes_symbols_NA_symbols_replaced_w_ENS_ids), genes_symbols_NA_symbols_replaced_w_ENS_ids$symbols)
   ###
     head(genes_symbols_NA_symbols_replaced_w_ENS_ids)
  symbols
  # ENSMUSG00000000001   Gnai3
  # ENSMUSG00000000028   Cdc45
  # ENSMUSG00000000037   Scml2
  # ENSMUSG00000000049    Apoh
  # ENSMUSG00000000056    Narf
  # ENSMUSG00000000058    Cav2

   table(is.na(genes_symbols_NA_symbols_replaced_w_ENS_ids$symbols))

  # FALSE
  # 25202
  #


  #### NOW, heatmap labels for genes are ready:

  #### DRG Top 50 up and Top 50 down genes:

  heatmap.2(rld_filtered_last_matrix[,1:5][c(DRG_F_sig_Log05_up_genes,DRG_F_sig_Log05_down_genes),], scale = "row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = genes_symbols_NA_symbols_replaced_w_ENS_ids[c(DRG_F_sig_Log05_up_genes,DRG_F_sig_Log05_down_genes),] ,margins = c(1, 16) ,dendrogram = "none", Rowv = NULL ,cexCol = 0.5, cexRow = 1, key = FALSE, ColSideColors = as.character(c("blue", "blue", "yellow", "yellow","yellow")), Colv = FALSE, labCol = FALSE)

  ### Nodose:

  heatmap.2(rld_filtered_last_matrix[,6:9][c(Nodose_F_sig_Log05_up_genes, Nodose_F_sig_Log05_down_genes),], scale = "row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = genes_symbols_NA_symbols_replaced_w_ENS_ids[c(Nodose_F_sig_Log05_up_genes, Nodose_F_sig_Log05_down_genes),] ,margins = c(1, 16) ,dendrogram = "none", Rowv = NULL ,cexCol = 0.5, cexRow = 1, key = FALSE, ColSideColors = as.character(c("blue", "blue", "yellow", "yellow")), labCol = FALSE,  Colv = FALSE)

  #####

  ### (Device size)-  8.5 x 16 exported.
  ### tiff: 722x1500


###########################################################################

  ### Sample Correlation Matrix:
  ####
  input_all_counts = counts(dds_drg_nod)
  library(psych)

  library(psych)
   correlation_matrix_raw_counts_pearson_BH = corr.test(input_all_counts, y = NULL, method = "pearson", adjust = "BH")
   correlation_matrix_raw_counts_pearson_BH$r
  # DRG_CS1_S4 DRG_CS2_S5  DRG_F_S2  DRG_F_S1  DRG_F_S3
  # DRG_CS1_S4  1.0000000  0.9884330 0.9464190 0.4768111 0.9735138
  # DRG_CS2_S5  0.9884330  1.0000000 0.9465326 0.5087419 0.9661598
  # DRG_F_S2    0.9464190  0.9465326 1.0000000 0.6719451 0.9745695
  # DRG_F_S1    0.4768111  0.5087419 0.6719451 1.0000000 0.5421805
  # DRG_F_S3    0.9735138  0.9661598 0.9745695 0.5421805 1.0000000
  # NG_CS_S8    0.9292836  0.9378726 0.9083089 0.5579856 0.9061801
  # NG_CS_S9    0.9124935  0.9254626 0.8830293 0.4929021 0.8903938
  # NG_F_S6     0.9336693  0.9404025 0.8859870 0.4539269 0.9085370
  # NG_F_S7     0.9359594  0.9337019 0.8895874 0.4444713 0.9173145
  # NG_CS_S8  NG_CS_S9   NG_F_S6   NG_F_S7
  # DRG_CS1_S4 0.9292836 0.9124935 0.9336693 0.9359594
  # DRG_CS2_S5 0.9378726 0.9254626 0.9404025 0.9337019
  # DRG_F_S2   0.9083089 0.8830293 0.8859870 0.8895874
  # DRG_F_S1   0.5579856 0.4929021 0.4539269 0.4444713
  # DRG_F_S3   0.9061801 0.8903938 0.9085370 0.9173145
  # NG_CS_S8   1.0000000 0.9691355 0.9675829 0.9644685
  # NG_CS_S9   0.9691355 1.0000000 0.9751286 0.9613528
  # NG_F_S6    0.9675829 0.9751286 1.0000000 0.9921442
  # NG_F_S7    0.9644685 0.9613528 0.9921442 1.0000000
  #
   CM = correlation_matrix_raw_counts_pearson_BH$r
   CP = correlation_matrix_raw_counts_pearson_BH$p

   library("corrplot")

   corrplot(CM, method="color", type="upper", addCoef.col = "black", tl.col="black", tl.srt=90, p.mat = CP, sig.level = 0.01, insig = "blank", diag=FALSE)


  ### EXPORTED 1000 x 1556 tiff.

   corrplot(CM, method="color", type="upper", addCoef.col = "black", tl.col="black", tl.srt=90, p.mat = CP, sig.level = 0.01, insig = "blank", diag=FALSE, order = "original", number.cex = 1.5)

   corrplot(CM, method="color", type="upper", addCoef.col = "black", tl.col="black", tl.srt=90, p.mat = CP, sig.level = 0.01, insig = "blank", diag=FALSE, order = "original", number.cex = 1.5, tl.cex = 1.25)

  ### parameters explained:

  # number.cex : coefficient font size.
  #
  # tl.cex : sample names font size.
  #
  # cl.cex : right color label font size.
  #
  # original: it means not to cluster so order is as given in the input matrix.

  corrplot(CM, method="color", type="upper", addCoef.col = "black", tl.col="black", tl.srt=90, p.mat = CP, sig.level = 0.01, insig = "blank", diag=FALSE, order = "original", number.cex = 1.5, tl.cex = 1.25, cl.cex = 1.25)

  ### it takes row.names and colnames from matrix. To change it:

   row.names(CM) = c("DRG-T1", "DRG-T2", "DRG-HS2", "DRG-HS1", "DRG-HS3","NG-T1", "NG-T2", "NG-HS1", "NG-HS-2")
   colnames(CM) = c("DRG-T1", "DRG-T2", "DRG-HS2", "DRG-HS1", "DRG-HS3","NG-T1", "NG-T2", "NG-HS1", "NG-HS-2")
   row.names(CP) = c("DRG-T1", "DRG-T2", "DRG-HS2", "DRG-HS1", "DRG-HS3","NG-T1", "NG-T2", "NG-HS1", "NG-HS-2")
   colnames(CP) = c("DRG-T1", "DRG-T2", "DRG-HS2", "DRG-HS1", "DRG-HS3","NG-T1", "NG-T2", "NG-HS1", "NG-HS-2")
   ####
   corrplot(CM, method="color", type="upper", addCoef.col = "black", tl.col="black", tl.srt=90, p.mat = CP, sig.level = 0.01, insig = "blank", diag=FALSE, order = "original", number.cex = 1.5, tl.cex = 1.25, cl.cex = 1.25)

  ##################################

  ### corrplot with filtered counts:

   dim(rld_filtered_last_matrix)
#  [1] 17843     9
   input_all_counts_filtered = input_all_counts[row.names(rld_filtered_last_matrix),]
   ###
   dim(input_all_counts_filtered)
#  [1] 17843     9


###
  correlation_matrix_filtered_counts_pearson_BH = corr.test(input_all_counts_filtered, y = NULL, method = "pearson", adjust = "BH")
  correlation_matrix_filtered_counts_pearson_BH$r


   CM_filtered = correlation_matrix_filtered_counts_pearson_BH$r
   CP_filtered = correlation_matrix_filtered_counts_pearson_BH$p
   row.names(CM_filtered) = c("DRG-T1", "DRG-T2", "DRG-HS2", "DRG-HS1", "DRG-HS3","NG-T1", "NG-T2", "NG-HS1", "NG-HS-2")

   colnames(CM_filtered) = c("DRG-T1", "DRG-T2", "DRG-HS2", "DRG-HS1", "DRG-HS3","NG-T1", "NG-T2", "NG-HS1", "NG-HS-2")

   row.names(CP_filtered) = c("DRG-T1", "DRG-T2", "DRG-HS2", "DRG-HS1", "DRG-HS3","NG-T1", "NG-T2", "NG-HS1", "NG-HS-2")

   colnames(CP_filtered) = c("DRG-T1", "DRG-T2", "DRG-HS2", "DRG-HS1", "DRG-HS3","NG-T1", "NG-T2", "NG-HS1", "NG-HS-2")

   corrplot(CM_filtered, method="color", type="upper", addCoef.col = "black", tl.col="black", tl.srt=90, p.mat = CP_filtered, sig.level = 0.01, insig = "blank", diag=FALSE, order = "original", number.cex = 1.5, tl.cex = 1.25, cl.cex = 1.25)

  ####### Two matrixes are almost identical.

####################################################################################################


   ### DRG ve NODOZ top 100 listesindeki ensembl id olan genler

   ## yukarıdan aşağı sıralı olarak tek tek ensembl websitesinden aratıldı.

   ### karşılıkları bulundu:

   # ## nodoz:
   #
   # Gm18562
   #
   # Gm5900
   #
   # Gm45713
   #
   # Gm5564
   #
   # Sptbn5
   #
   # ### DRG:
   #
   # Gm44821
   #
   # Gm10031
   #
   # Gm28438
   #
   # Gm5900
   #
   # Gm10925
   #
   # Gm8995
   #
   # ### Bunlar top 100 genlere eklenip tekrar heatmap yapılacak:

    genes_symbols_NA_symbols_replaced_w_ENS_ids[c(DRG_F_sig_Log05_up_genes,DRG_F_sig_Log05_down_genes),]



   # [1] "Cysltr2"            "Tnfaip8"            "Nlrp5-ps"           "Zfp760"
   # [5] "ENSMUSG00000109540" "COX3"               "Adk"                "Ap3s1"
   # [9] "Zfp52"              "Nppb"               "Nae1"               "ENSMUSG00000101523"
   # [13] "Apoo"               "ENSMUSG00000101939" "Pde4c"              "Rab27b"
   # [17] "Far2"               "Ppef1"              "Xpo1"               "Cd55"
   # [21] "Hbb-b1"             "Gmfb"               "Auts2"              "Zmym1"
   # [25] "Smyd3"              "Mrgpra3"            "ENSMUSG00000094685" "Kcnip4"
   # [29] "Zfp942"             "Rgs4"               "Plcl2"              "COX1"
   # [33] "Ms4a3"              "Rgs17"              "ENSMUSG00000100862" "Nptx2"
   # [37] "Kpna4"              "Pdia6"              "Mal2"               "Ptges3"
   # [41] "Mrgprx1"            "Amd1"               "Tmx3"               "Klhl7"
   # [45] "Cpne3"              "Vstm5"              "Dpp10"              "Ormdl1"
   # [49] "Lxn"                "Cd47"               "Kdr"                "Flt1"
   # [53] "Slco1a4"            "Igfbp7"             "Ctla2a"             "Cd93"
   # [57] "Tm4sf1"             "Pecam1"             "Cldn5"              "Adgrl4"
   # [61] "Cavin2"             "S100a9"             "Gimap6"             "Slfn5"
   # [65] "Prom1"              "Myh11"              "Abcb1a"             "Sox17"
   # [69] "Icam1"              "Acvrl1"             "Dcn"                "Apold1"
   # [73] "Clic5"              "Smoc2"              "Ly6c1"              "Emcn"
   # [77] "Galnt15"            "C130074G19Rik"      "Akr1c14"            "Sgk1"
   # [81] "Tgm2"               "Rgs16"              "Dll4"               "Sele"
   # [85] "Tek"                "ENSMUSG00000063286" "S100a8"             "Gsn"
   # [89] "Gbp4"               "Spry1"              "Tie1"               "Tmem204"
   # [93] "Tgtp1"              "Srgn"               "Igfbp6"             "Cxcl12"
   # [97] "Clec2d"             "Esam"               "Podxl"              "Mgp"
   #

   genes_symbols_NA_symbols_replaced_w_ENS_ids[c(Nodose_F_sig_Log05_up_genes, Nodose_F_sig_Log05_down_genes),]
   # [1] "ENSMUSG00000110123" "Fgf3"               "Lin7a"              "Mum1l1"
   # [5] "Ttc38"              "Man2a1"             "Ahsa1"              "Plcl2"
   # [9] "Ahcy"               "ENSMUSG00000094685" "Far2"               "Slc25a40"
   # [13] "Ahsa2"              "Nlrp5-ps"           "Mkx"                "Lypd6"
   # [17] "Tango2"             "Thoc6"              "Wdr75"              "P4ha2"
   # [21] "Mrpl12"             "Pycrl"              "Serpini1"           "Cacybp"
   # [25] "Hdhd2"              "Intu"               "ENSMUSG00000089989" "Spsb4"
   # [29] "Fam69b"             "Nhej1"              "Bbs12"              "Vav3"
   # [33] "Amacr"              "Coq4"               "Ptprr"              "Uqcc1"
   # [37] "Mettl16"            "Nebl"               "ENSMUSG00000046440" "Ttll9"
   # [41] "Gap43"              "Nme1"               "Gal"                "Aox3"
   # [45] "Ikzf3"              "Rab39"              "Tomm40l"            "Tusc2"
   # [49] "Telo2"              "Fkbp4"              "Prx"                "Cav1"
   # [53] "Dcn"                "Ncmap"              "Slco1a4"            "Pecam1"
   # [57] "Mbp"                "Cd93"               "Cldn5"              "Myh11"
   # [61] "Smoc2"              "Cldn19"             "Arhgap19"           "Acvrl1"
   # [65] "Apod"               "Esam"               "Ly6c1"              "Ogn"
   # [69] "Sik1"               "Slc36a2"            "Sgk1"               "Sox17"
   # [73] "Emp1"               "Efhd1"              "Il16"               "Tek"
   # [77] "Fam178b"            "Tcim"               "Mal"                "Itga6"
   # [81] "Ccl11"              "Txnip"              "Bcas1"              "Shroom4"
   # [85] "Slfn5"              "Fli1"               "Igfbp6"             "Kcna1"
   # [89] "Drp2"               "Col5a1"             "Lyz2"               "ENSMUSG00000074899"
   # [93] "Mpz"                "Pdgfra"             "Ptprb"              "Eps8l2"
   # [97] "Ecscr"              "Pxn"                "Reln"               "Nes"

    nodose_top_100_up_down_symbols = genes_symbols_NA_symbols_replaced_w_ENS_ids[c(Nodose_F_sig_Log05_up_genes, Nodose_F_sig_Log05_down_genes),]
    drg_top_100_up_down_symbols = genes_symbols_NA_symbols_replaced_w_ENS_ids[c(DRG_F_sig_Log05_up_genes,DRG_F_sig_Log05_down_genes),]

    nodose_top_100_up_down_symbols

    drg_top_100_up_down_symbols

    nodose_top_100_up_down_symbols[1]
#   [1] "ENSMUSG00000110123"

    nodose_top_100_up_down_symbols[1] = "Gm18562"
    nodose_top_100_up_down_symbols[10] = "Gm5900"
    nodose_top_100_up_down_symbols[27] = "Gm45713"
    nodose_top_100_up_down_symbols[39] = "Gm5564"
    nodose_top_100_up_down_symbols[92] = "Sptbn5"

    nodose_top_100_up_down_symbols

   # [1] "Gm18562"  "Fgf3"     "Lin7a"    "Mum1l1"   "Ttc38"    "Man2a1"   "Ahsa1"    "Plcl2"
   # [9] "Ahcy"     "Gm5900"   "Far2"     "Slc25a40" "Ahsa2"    "Nlrp5-ps" "Mkx"      "Lypd6"
   # [17] "Tango2"   "Thoc6"    "Wdr75"    "P4ha2"    "Mrpl12"   "Pycrl"    "Serpini1" "Cacybp"
   # [25] "Hdhd2"    "Intu"     "Gm45713"  "Spsb4"    "Fam69b"   "Nhej1"    "Bbs12"    "Vav3"
   # [33] "Amacr"    "Coq4"     "Ptprr"    "Uqcc1"    "Mettl16"  "Nebl"     "Gm5564"   "Ttll9"
   # [41] "Gap43"    "Nme1"     "Gal"      "Aox3"     "Ikzf3"    "Rab39"    "Tomm40l"  "Tusc2"
   # [49] "Telo2"    "Fkbp4"    "Prx"      "Cav1"     "Dcn"      "Ncmap"    "Slco1a4"  "Pecam1"
   # [57] "Mbp"      "Cd93"     "Cldn5"    "Myh11"    "Smoc2"    "Cldn19"   "Arhgap19" "Acvrl1"
   # [65] "Apod"     "Esam"     "Ly6c1"    "Ogn"      "Sik1"     "Slc36a2"  "Sgk1"     "Sox17"
   # [73] "Emp1"     "Efhd1"    "Il16"     "Tek"      "Fam178b"  "Tcim"     "Mal"      "Itga6"
   # [81] "Ccl11"    "Txnip"    "Bcas1"    "Shroom4"  "Slfn5"    "Fli1"     "Igfbp6"   "Kcna1"
   # [89] "Drp2"     "Col5a1"   "Lyz2"     "Sptbn5"   "Mpz"      "Pdgfra"   "Ptprb"    "Eps8l2"
   # [97] "Ecscr"    "Pxn"      "Reln"     "Nes"

     drg_top_100_up_down_symbols
   # [1] "Cysltr2"            "Tnfaip8"            "Nlrp5-ps"           "Zfp760"
   # [5] "ENSMUSG00000109540" "COX3"               "Adk"                "Ap3s1"
   # [9] "Zfp52"              "Nppb"               "Nae1"               "ENSMUSG00000101523"
   # [13] "Apoo"               "ENSMUSG00000101939" "Pde4c"              "Rab27b"
   # [17] "Far2"               "Ppef1"              "Xpo1"               "Cd55"
   # [21] "Hbb-b1"             "Gmfb"               "Auts2"              "Zmym1"
   # [25] "Smyd3"              "Mrgpra3"            "ENSMUSG00000094685" "Kcnip4"
   # [29] "Zfp942"             "Rgs4"               "Plcl2"              "COX1"
   # [33] "Ms4a3"              "Rgs17"              "ENSMUSG00000100862" "Nptx2"
   # [37] "Kpna4"              "Pdia6"              "Mal2"               "Ptges3"
   # [41] "Mrgprx1"            "Amd1"               "Tmx3"               "Klhl7"
   # [45] "Cpne3"              "Vstm5"              "Dpp10"              "Ormdl1"
   # [49] "Lxn"                "Cd47"               "Kdr"                "Flt1"
   # [53] "Slco1a4"            "Igfbp7"             "Ctla2a"             "Cd93"
   # [57] "Tm4sf1"             "Pecam1"             "Cldn5"              "Adgrl4"
   # [61] "Cavin2"             "S100a9"             "Gimap6"             "Slfn5"
   # [65] "Prom1"              "Myh11"              "Abcb1a"             "Sox17"
   # [69] "Icam1"              "Acvrl1"             "Dcn"                "Apold1"
   # [73] "Clic5"              "Smoc2"              "Ly6c1"              "Emcn"
   # [77] "Galnt15"            "C130074G19Rik"      "Akr1c14"            "Sgk1"
   # [81] "Tgm2"               "Rgs16"              "Dll4"               "Sele"
   # [85] "Tek"                "ENSMUSG00000063286" "S100a8"             "Gsn"
   # [89] "Gbp4"               "Spry1"              "Tie1"               "Tmem204"
   # [93] "Tgtp1"              "Srgn"               "Igfbp6"             "Cxcl12"
   # [97] "Clec2d"             "Esam"               "Podxl"              "Mgp"
    drg_top_100_up_down_symbols[5]
#   [1] "ENSMUSG00000109540"
    drg_top_100_up_down_symbols[5] = "Gm44821"
    drg_top_100_up_down_symbols[12] = "Gm10031"
    drg_top_100_up_down_symbols[14] = "Gm28438"
    drg_top_100_up_down_symbols[27] = "Gm5900"
    drg_top_100_up_down_symbols[35]
#   [1] "ENSMUSG00000100862"
    drg_top_100_up_down_symbols[35] = "Gm10925"
    drg_top_100_up_down_symbols[86] = "Gm8995"
    ####
     drg_top_100_up_down_symbols
   # [1] "Cysltr2"       "Tnfaip8"       "Nlrp5-ps"      "Zfp760"        "Gm44821"
   # [6] "COX3"          "Adk"           "Ap3s1"         "Zfp52"         "Nppb"
   # [11] "Nae1"          "Gm10031"       "Apoo"          "Gm28438"       "Pde4c"
   # [16] "Rab27b"        "Far2"          "Ppef1"         "Xpo1"          "Cd55"
   # [21] "Hbb-b1"        "Gmfb"          "Auts2"         "Zmym1"         "Smyd3"
   # [26] "Mrgpra3"       "Gm5900"        "Kcnip4"        "Zfp942"        "Rgs4"
   # [31] "Plcl2"         "COX1"          "Ms4a3"         "Rgs17"         "Gm10925"
   # [36] "Nptx2"         "Kpna4"         "Pdia6"         "Mal2"          "Ptges3"
   # [41] "Mrgprx1"       "Amd1"          "Tmx3"          "Klhl7"         "Cpne3"
   # [46] "Vstm5"         "Dpp10"         "Ormdl1"        "Lxn"           "Cd47"
   # [51] "Kdr"           "Flt1"          "Slco1a4"       "Igfbp7"        "Ctla2a"
   # [56] "Cd93"          "Tm4sf1"        "Pecam1"        "Cldn5"         "Adgrl4"
   # [61] "Cavin2"        "S100a9"        "Gimap6"        "Slfn5"         "Prom1"
   # [66] "Myh11"         "Abcb1a"        "Sox17"         "Icam1"         "Acvrl1"
   # [71] "Dcn"           "Apold1"        "Clic5"         "Smoc2"         "Ly6c1"
   # [76] "Emcn"          "Galnt15"       "C130074G19Rik" "Akr1c14"       "Sgk1"
   # [81] "Tgm2"          "Rgs16"         "Dll4"          "Sele"          "Tek"
   # [86] "Gm8995"        "S100a8"        "Gsn"           "Gbp4"          "Spry1"
   # [91] "Tie1"          "Tmem204"       "Tgtp1"         "Srgn"          "Igfbp6"
   # [96] "Cxcl12"        "Clec2d"        "Esam"          "Podxl"         "Mgp"

   #####
   heatmap.2(rld_filtered_last_matrix[,1:5][c(DRG_F_sig_Log05_up_genes,DRG_F_sig_Log05_down_genes),], scale = "row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = drg_top_100_up_down_symbols ,margins = c(1, 16) ,dendrogram = "none", Rowv = NULL ,cexCol = 0.5, cexRow = 1, key = FALSE, ColSideColors = as.character(c("blue", "blue", "yellow", "yellow","yellow")), Colv = FALSE, labCol = FALSE)

    heatmap.2(rld_filtered_last_matrix[,6:9][c(Nodose_F_sig_Log05_up_genes, Nodose_F_sig_Log05_down_genes),], scale = "row", trace="none", col = colorpanel(64,"red", "black", "green"), labRow = nodose_top_100_up_down_symbols ,margins = c(1, 16) ,dendrogram = "none", Rowv = NULL ,cexCol = 0.5, cexRow = 1, key = FALSE, ColSideColors = as.character(c("blue", "blue", "yellow", "yellow")), labCol = FALSE,  Colv = FALSE)

   ## EXPORTED:

   ### (Device size)-  8.5 x 16 exported.
   ### tiff: 722x1500
####################################################################################################

    ### EXPORT FPKM Matrices for specific gene sets:

     write.csv(selected_TRP_genes ,file = "selected_TRP_genes.txt")
     write.csv(selected_purinergic_receptor_genes ,file = "selected_purinergic_receptor_genes.txt")
     write.csv(selected_mechano_genes ,file = "selected_mechano_genes.txt")
     write.csv(selected_channel_genes ,file = "selected_channel_genes.txt")
     write.csv(selected_gpr_genes ,file = "selected_gpr_genes.txt")
     write.csv(selected_G_protein_coupled_genes ,file = "selected_G_protein_coupled_genes.txt")

     dim(fpkm_drg_nod_matrix_w_padj)
    # [1] 17832    13

     fpkm_DRG_05_UP_DOWN_GPCR_genes = fpkm_drg_nod_matrix_w_padj[c(DRG_05_UP_GPCR_genes,DRG_05_DOWN_GPCR_genes),]
     write.csv(fpkm_DRG_05_UP_DOWN_GPCR_genes ,file = "fpkm_DRG_05_UP_DOWN_GPCR_genes.txt")
     fpkm_DRG_05_UP_DOWN_ion_channel_genes = fpkm_drg_nod_matrix_w_padj[c(DRG_05_UP_ion_channel_genes,DRG_05_DOWN_ion_channel_genes),]
     fpkm_DRG_05_UP_DOWN_TFactor_genes = fpkm_drg_nod_matrix_w_padj[c(DRG_05_UP_TFactor_genes,DRG_05_DOWN_TFactor_genes),]
     ####
     fpkm_Nodose_05_UP_DOWN_GPCR_genes = fpkm_drg_nod_matrix_w_padj[c(Nodose_05_UP_GPCR_genes,Nodose_05_DOWN_GPCR_genes),]
     fpkm_Nodose_05_UP_DOWN_ion_channel_genes = fpkm_drg_nod_matrix_w_padj[c(Nodose_05_UP_ion_channel_genes,Nodose_05_DOWN_ion_channel_genes),]
     fpkm_Nodose_05_UP_DOWN_TFactor_genes = fpkm_drg_nod_matrix_w_padj[c(Nodose_05_UP_TFactor_genes,Nodose_05_DOWN_TFactor_genes),]

########################################################################################################

#### BAR CHARTS:


      bar_chart_df[3,3] = "DRG-HS_vs_DRG-T"
      bar_chart_df[5,3] = "DRG-HS_vs_DRG-T"
      bar_chart_df[4,3] =  "NG-HS_vs_NG-T"
      bar_chart_df[6,3] =  "NG-HS_vs_NG-T"

      bar_chart_df

     # counts       type          sample
     # 1    238         UP DRG-HS_vs_DRG-T
     # 2     NA         UP   NG-HS_vs_NG-T
     # 3     NA       DOWN DRG-HS_vs_DRG-T
     # 4     NA       DOWN   NG-HS_vs_NG-T
     # 5     NA COMPARABLE DRG-HS_vs_DRG-T
     # 6     NA COMPARABLE   NG-HS_vs_NG-T



      bar_chart_df[2,1] =   Nodose_up_05_count
      bar_chart_df[4,1] =   Nodose_down_05_count
      bar_chart_df[6,1] =   NODOSE_all_genes_count

       bar_chart_df


     # counts       type          sample
     # 1    238         UP DRG-HS_vs_DRG-T
     # 2     66         UP   NG-HS_vs_NG-T
     # 3     NA       DOWN DRG-HS_vs_DRG-T
     # 4    761       DOWN   NG-HS_vs_NG-T
     # 5     NA COMPARABLE DRG-HS_vs_DRG-T
     # 6  17867 COMPARABLE   NG-HS_vs_NG-T



     bar_chart_df[5,1] =  DRG_all_genes_count


     bar_chart_df[3,1] = DRG_down_05_count


     bar_chart_df

     # counts       type          sample
     # 1    238         UP DRG-HS_vs_DRG-T
     # 2     66         UP   NG-HS_vs_NG-T
     # 3    757       DOWN DRG-HS_vs_DRG-T
     # 4    761       DOWN   NG-HS_vs_NG-T
     # 5  17867 COMPARABLE DRG-HS_vs_DRG-T
     # 6  17867 COMPARABLE   NG-HS_vs_NG-T


     bar_chart_df$counts[5]-(bar_chart_df$counts[1]+bar_chart_df$counts[3])

     #[1] 16872


     bar_chart_df$counts[6]-(bar_chart_df$counts[2]+bar_chart_df$counts[4])
     #[1] 17040

     ####
     bar_chart_df_1 = bar_chart_df
     bar_chart_df$type[5] = "TOTAL"
     bar_chart_df$type[6] = "TOTAL"

     bar_chart_df

     # counts  type          sample
     # 1    238    UP DRG-HS_vs_DRG-T
     # 2     66    UP   NG-HS_vs_NG-T
     # 3    757  DOWN DRG-HS_vs_DRG-T
     # 4    761  DOWN   NG-HS_vs_NG-T
     # 5  17867 TOTAL DRG-HS_vs_DRG-T
     # 6  17867 TOTAL   NG-HS_vs_NG-T

     bar_chart_df_1$counts[5] = 16872

     bar_chart_df_1$counts[6] = 17040

     bar_chart_df_1

     # counts       type          sample
     # 1    238         UP DRG-HS_vs_DRG-T
     # 2     66         UP   NG-HS_vs_NG-T
     # 3    757       DOWN DRG-HS_vs_DRG-T
     # 4    761       DOWN   NG-HS_vs_NG-T
     # 5  16872 COMPARABLE DRG-HS_vs_DRG-T
     # 6  17040 COMPARABLE   NG-HS_vs_NG-T

     colnames(bar_chart_df_1)[1] = "genes"

     # genes       type          sample
     # 1   238         UP DRG-HS_vs_DRG-T
     # 2    66         UP   NG-HS_vs_NG-T
     # 3   757       DOWN DRG-HS_vs_DRG-T
     # 4   761       DOWN   NG-HS_vs_NG-T
     # 5 16872 COMPARABLE DRG-HS_vs_DRG-T
     # 6 17040 COMPARABLE   NG-HS_vs_NG-T

     ## all types:

     ggplot2.barplot(data= bar_chart_df_1, xName="sample", yName = "genes" , groupName="type", groupColors = c("blue","green", "red"))

     ### only up and down genes:

     ggplot2.barplot(data= head(bar_chart_df_1, 4), xName="sample", yName = "genes" , groupName="type", groupColors = c("blue","red"))

     ### x eksenini alfabetik sıralıyor.

###########################################################################################################

     ### Venn Diagrams:

     biocLite("VennDiagram")
     require("VennDiagram")
     drg_venn = c(DRG_sig_all_log05_up_genes,DRG_sig_all_log05_down_genes)
     nodose_venn = c(Nodose_sig_all_log05_up_genes,Nodose_sig_all_log05_down_genes)
     venn.diagram(list("DRG-HS" = drg_venn, "NODOSE-HS" = nodose_venn),fill = c("blue", "red"), alpha = c(0.5, 0.5), cex = 2,cat.fontface = 4, filename = "venn_plot_1")

     ### without names:

     venn.diagram(list(" " = drg_venn, " " = nodose_venn),fill = c("blue", "red"), alpha = c(0.5, 0.5), cex = 2,cat.fontface = 4, filename = "venn_plot_2")

     ###########

      DRG_NDS_up_intersect = intersect(DRG_sig_all_log05_up_genes,Nodose_sig_all_log05_up_genes)
      DRG_NDS_down_intersect = intersect(DRG_sig_all_log05_down_genes,Nodose_sig_all_log05_down_genes)

      length(DRG_NDS_up_intersect)
     # [1] 10
      length(DRG_NDS_down_intersect)
     # [1] 335
      length(DRG_sig_all_log05_up_genes)
     # [1] 238
      length(DRG_sig_all_log05_down_genes)
     # [1] 757
      length(Nodose_sig_all_log05_up_genes)
     # [1] 66
      length(Nodose_sig_all_log05_down_genes)
     # [1] 761

      ###
       DRG_sig_all_log05_genes = c(DRG_sig_all_log05_up_genes,DRG_sig_all_log05_down_genes)
      Nodose_sig_all_log05_genes = c(Nodose_sig_all_log05_up_genes,Nodose_sig_all_log05_down_genes)
      ###
       length(DRG_sig_all_log05_genes)
     # [1] 995
      length(Nodose_sig_all_log05_genes)
     # [1] 827

     ########

      fpkm_DRG_sig_all_log05_up_genes = fpkm_drg_nod_matrix_w_padj[DRG_sig_all_log05_up_genes,]
      fpkm_DRG_sig_all_log05_down_genes = fpkm_drg_nod_matrix_w_padj[DRG_sig_all_log05_down_genes,]
      fpkm_DRG_sig_all_log05_genes = fpkm_drg_nod_matrix_w_padj[DRG_sig_all_log05_genes,]

     ### Not: fpkm matriksi DRG'de ekspres olan "ENSMUSG00000052305" genini
     # içermiyor. O yüzden o gen için "NA" row verdi. Manuel olarak excelde o genin
     # fpkm bilgisini girdik.

      fpkm_Nodose_sig_all_log05_genes = fpkm_drg_nod_matrix_w_padj[Nodose_sig_all_log05_genes,]
      fpkm_Nodose_sig_all_log05_up_genes = fpkm_drg_nod_matrix_w_padj[Nodose_sig_all_log05_up_genes,]
      fpkm_Nodose_sig_all_log05_down_genes = fpkm_drg_nod_matrix_w_padj[Nodose_sig_all_log05_down_genes,]

      fpkm_DRG_NDS_up_intersect = fpkm_drg_nod_matrix_w_padj[DRG_NDS_up_intersect,]
      fpkm_DRG_NDS_down_intersect = fpkm_drg_nod_matrix_w_padj[DRG_NDS_down_intersect,]

      ####
      write.csv(fpkm_DRG_sig_all_log05_up_genes, file = "fpkm_DRG_sig_all_log05_up_genes.txt")
      write.csv( fpkm_DRG_sig_all_log05_down_genes, file = "fpkm_DRG_sig_all_log05_down_genes.txt")
      write.csv(fpkm_DRG_sig_all_log05_genes, file = "fpkm_DRG_sig_all_log05_genes.txt")
      write.csv(fpkm_Nodose_sig_all_log05_genes, file = "fpkm_Nodose_sig_all_log05_genes.txt")
      write.csv(fpkm_Nodose_sig_all_log05_up_genes, file = "fpkm_Nodose_sig_all_log05_up_genes.txt")
      write.csv(fpkm_Nodose_sig_all_log05_down_genes, file = "fpkm_Nodose_sig_all_log05_down_genes.txt")
      write.csv(fpkm_DRG_NDS_up_intersect, file = "fpkm_DRG_NDS_up_intersect.txt")
      write.csv(fpkm_DRG_NDS_down_intersect, file = "fpkm_DRG_NDS_down_intersect.txt")


##########################################################################################################


# GENE ABUNDANCES AS TPM VALUES


      ### Tximport is installed:

       if (!requireNamespace("BiocManager", quietly = TRUE))
        + install.packages("BiocManager")

       BiocManager::install("tximport", version = "3.8")

       library(tximport)

      ### According to Mike Love: "tx.salmon$abundance will be TPM table from the original salmon quantification, but gene-level summarized."

      ### So to obtain TPM matrix:

      head(txi_mouse_ens_91_drg_nod$abundance)

      # DRG_CS1_S4 DRG_CS2_S5  DRG_F_S2  DRG_F_S1  DRG_F_S3  NG_CS_S8  NG_CS_S9   NG_F_S6
      # ENSMUSG00000000001  63.366957  60.244832 67.190185 64.381081 70.225357 54.403782 49.987600 47.255478
      # ENSMUSG00000000003   0.000000   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000
      # ENSMUSG00000000028   4.286111   3.717176  3.283623  2.937711  3.117223  3.619425  1.878272  2.733708
      # ENSMUSG00000000037   0.029050   0.080325  0.034216  0.042651  0.000000  0.152270  0.389782  0.082705
      # ENSMUSG00000000049   3.020117   3.012096  3.055215  4.786893  4.453607  4.523055  5.052050  4.664443
      # ENSMUSG00000000056  18.281877  18.903672 23.285522 12.457277 21.052795 12.610299 10.445226 18.056432
      # NG_F_S7
      # ENSMUSG00000000001 49.834000
      # ENSMUSG00000000003  0.000000
      # ENSMUSG00000000028  2.922635
      # ENSMUSG00000000037  0.021711
      # ENSMUSG00000000049  5.103724
      # ENSMUSG00000000056 15.954042


       length(row.names(txi_mouse_ens_91_drg_nod$abundance))
      # [1] 34517

       txi_mouse_ens_91_drg_nod_abundance_DF = as.data.frame(txi_mouse_ens_91_drg_nod$abundance)

      ### "rld_drg_nod" includes names of genes, whose raw-vise sum of count numbers > 1".
      ### we can filter our TPM list, only keeping these gene names:

       txi_mouse_ens_91_drg_nod_abundance_DF_filtered = txi_mouse_ens_91_drg_nod_abundance_DF[row.names(rld_drg_nod),]


       write.csv(txi_mouse_ens_91_drg_nod_abundance_DF, file = ("txi_mouse_ens_91_drg_nod_abundance_DF.txt"))


       write.csv(txi_mouse_ens_91_drg_nod_abundance_DF_filtered, file = ("txi_mouse_ens_91_drg_nod_abundance_DF_filtered.txt"))




##########################################################################################################

       # TPM Lists For EPEE TOOL:


       ### filtered TPM list will be prepared from TPM lists.

       #### results with "NA" and duplicated symbols removed will be used for filtering.

       #### EPEE wants control and treatment samples in separate tab-delimited files, with gene symbol raw names.

       #### We will prepare TPM lists and export them as .csv files.

       ### Final lists for EPEE will be preapared on EXCEL.







       library(DESeq2)

        TPM_DRG_Na_omit_duplicate_symbol_rm = txi_mouse_ens_91_drg_nod_abundance_DF_filtered[row.names(res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup),]
        TPM_Nodose_Na_omit_duplicate_symbol_rm = txi_mouse_ens_91_drg_nod_abundance_DF_filtered[row.names(res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup),]
        ####
        TPM_DRG_Na_omit_duplicate_symbol_rm$symbols = res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup$symbols
        TPM_Nodose_Na_omit_duplicate_symbol_rm$symbols = res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup$symbols
        ####
       head(TPM_DRG_Na_omit_duplicate_symbol_rm)


       head(TPM_Nodose_Na_omit_duplicate_symbol_rm)

       ### row.names are altered to symbols:
        TPM_DRG_Na_omit_with_unique_symbol_raw_names = TPM_DRG_Na_omit_duplicate_symbol_rm
        row.names(TPM_DRG_Na_omit_with_unique_symbol_raw_names) = TPM_DRG_Na_omit_with_unique_symbol_raw_names$symbols
       ####
       TPM_Nodose_Na_omit_with_unique_symbol_raw_names = TPM_Nodose_Na_omit_duplicate_symbol_rm
       row.names(TPM_Nodose_Na_omit_with_unique_symbol_raw_names) = TPM_Nodose_Na_omit_with_unique_symbol_raw_names$symbols
       ####
       head(TPM_DRG_Na_omit_with_unique_symbol_raw_names)

       head(TPM_Nodose_Na_omit_with_unique_symbol_raw_names)

        ### Export:
       write.csv(TPM_DRG_Na_omit_with_unique_symbol_raw_names, file = "TPM_DRG_Na_omit_with_unique_symbol_raw_names")
       write.csv(TPM_Nodose_Na_omit_with_unique_symbol_raw_names, file = "TPM_Nodose_Na_omit_with_unique_symbol_raw_names")

##########################################################################################################

       #### Diff. Exp. analysis of DRG_F vs Nodose_F:

        library(DESeq2)


         res_DRG_F_vs_Nodose_F = results(dds_drg_nod, contrast = c("condition", "sample_drg","sample_nodose"), alpha = 0.05)

        #### Filterings:

        res_DRG_F_vs_Nodose_F_na_omit = na.omit(res_DRG_F_vs_Nodose_F)

        ## OPTIONAL: "anyNA" function can be used as follows:

         anyNA(res_DRG_F_vs_Nodose_F_na_omit, recursive=FALSE)

#     [1] FALSE

       table(duplicated(res_DRG_F_vs_Nodose_F_na_omit$symbols))

       ### We will order by padj values:
        res_DRG_F_vs_Nodose_F_na_omit = res_DRG_F_vs_Nodose_F_na_omit[order( res_DRG_F_vs_Nodose_F_na_omit$padj),]

       ### we will replace "NA" symbols with ensembl IDs.

        #### Adding gene definitions:
         res_DRG_F_vs_Nodose_F_na_omit$gene_names = mapIds(org.Mm.eg.db, keys = row.names(res_DRG_F_vs_Nodose_F_na_omit), keytype = "ENSEMBL", column = "GENENAME", multiVals = "first" )

        #### EXPORT files:

         write.csv(res_DRG_F_vs_Nodose_F_na_omit, file = "res_DRG_F_vs_Nodose_F_na_omit.txt")




##########################################################################################################


         ### Let's take the Nodose samples' count matrix:

         library(DESeq2)


          dds_drg_nod_unfiltered_counts = DESeqDataSetFromTximport(txi_mouse_ens_91_drg_nod,sample_table_drg_ng, ~condition)

        # using counts and average transcript lengths from tximport

          dds_drg_nod_unfiltered_counts

         # class: DESeqDataSet
         # dim: 34517 9
         # metadata(1): version
         # assays(2): counts avgTxLength
         # rownames(34517): ENSMUSG00000000001 ENSMUSG00000000003 ... ENSMUSG00000115847
         # ENSMUSG00000115848
         # rowData names(0):
         #   colnames(9): DRG_CS1_S4 DRG_CS2_S5 ... NG_F_S6 NG_F_S7
         # colData names(2): SampleNames condition

          unfiltered_counts = counts(dds_drg_nod_unfiltered_counts)
          head(unfiltered_counts)

          head(unfiltered_counts[,6:9])

          nodose_unfiltered_counts = unfiltered_counts[,6:9]
          dim(nodose_unfiltered_counts)
         # [1] 34517     4
          write.csv(nodose_unfiltered_counts, file = "nodose_unfiltered_counts.txt")


##########################################################################################################

            res_DRG_F_vs_cont_drg_symbols_NA_keeped_sig_log_15_DOWN = subset(res_DRG_F_vs_cont_drg_symbols_NA_keeped_sig_log05_DOWN, log2FoldChange < -1.4999999999)

           length(res_DRG_F_vs_cont_drg_symbols_NA_keeped_sig_log_15_DOWN$baseMean)

            # [1] 384


          genes_ids_drg_down__from_NA_keeped = as.character(row.names(res_DRG_F_vs_cont_drg_symbols_NA_keeped_sig_log_15_DOWN))

          res_DRG_F_vs_cont_drg_symbols_NA_keeped_sig_log_15_DOWN

          fpkm_drg_down_from_NA_keeped= fpkm_drg_nod_matrix_w_padj[genes_ids_drg_down__from_NA_keeped,]

          # head(fpkm_drg_down_from_NA_keeped)


          ### Combine different columns with same rows, with "cbind" function:

           fpkm_drg_down_from_NA_keeped_for_GURKAN_HOCA = cbind(fpkm_drg_down_from_NA_keeped[,1:5], fpkm_drg_down_from_NA_keeped[,10:12])

           head(fpkm_drg_down_from_NA_keeped_for_GURKAN_HOCA)

          #  RG_CS1_S4 DRG_CS2_S5  DRG_F_S2  DRG_F_S1  DRG_F_S3 symbols                                                   gene_name     DRG_Padj
          # ENSMUSG00000062960   4.720781   5.310751 0.2046565 0.2885878 0.1959173     Kdr                       kinase insert domain protein receptor 2.180806e-70
          # ENSMUSG00000029648   8.893749   7.769431 0.1512997 0.2048175 0.1932051    Flt1                                  FMS-like tyrosine kinase 1 2.918220e-63
          # ENSMUSG00000030237  15.918621  10.817824 0.2962841 0.2260122 0.1557420 Slco1a4 solute carrier organic anion transporter family, member 1a4 2.918220e-63
          # ENSMUSG00000036256  71.439000 106.417048 2.7152044 3.6766615 2.0942917  Igfbp7                insulin-like growth factor binding protein 7 7.371914e-62
          # ENSMUSG00000044258  29.072261  34.471836 1.4518042 1.5547670 1.1324595  Ctla2a           cytotoxic T lymphocyte-associated protein 2 alpha 2.642365e-60
          # ENSMUSG00000027435   5.692715   9.004278 0.1956998 0.2708483 0.1442096    Cd93                                                CD93 antigen 1.862927e-59
          #

           dim(fpkm_drg_down_from_NA_keeped_for_GURKAN_HOCA)
          # [1] 384   8

          ### Add log2foldchange values:

           fpkm_drg_down_from_NA_keeped_for_GURKAN_HOCA[,9] = res_DRG_F_vs_cont_drg_symbols_NA_keeped_sig_log_15_DOWN$log2FoldChange

           head(fpkm_drg_down_from_NA_keeped_for_GURKAN_HOCA)



          colnames(fpkm_drg_down_from_NA_keeped_for_GURKAN_HOCA)
          # [1] "DRG_CS1_S4" "DRG_CS2_S5" "DRG_F_S2"   "DRG_F_S1"   "DRG_F_S3"   "symbols"    "gene_name"  "DRG_Padj"   "V9"

           colnames(fpkm_drg_down_from_NA_keeped_for_GURKAN_HOCA)[9]
          # [1] "V9"

          ### Change the name of the last column:

           colnames(fpkm_drg_down_from_NA_keeped_for_GURKAN_HOCA)[9] = "log2FoldChange"
           colnames(fpkm_drg_down_from_NA_keeped_for_GURKAN_HOCA)
          # [1] "DRG_CS1_S4"     "DRG_CS2_S5"     "DRG_F_S2"       "DRG_F_S1"       "DRG_F_S3"       "symbols"        "gene_name"      "DRG_Padj"       "log2FoldChange"

          ### Add also thebasemean values as a new column:

           fpkm_drg_down_from_NA_keeped_for_GURKAN_HOCA$baseMean = res_DRG_F_vs_cont_drg_symbols_NA_keeped_sig_log_15_DOWN$baseMean

           write.csv(fpkm_drg_down_from_NA_keeped_for_GURKAN_HOCA, file = "fpkm_drg_down_from_NA_keeped_for_GURKAN_HOCA.txt")




##########################################################################################################

   ### We will create MA Plots annotated with select gene names.


   ### Check row.names type of dds object:

   dds_drg_nod

   # class: DESeqDataSet
   # dim: 25202 9
   # metadata(1): version
   # assays(5): counts avgTxLength normalizationFactors mu
   # cooks
   # rownames(25202): ENSMUSG00000000001 ENSMUSG00000000028
   # ... ENSMUSG00000115847 ENSMUSG00000115848
   # rowData names(29): baseMean baseVar ... deviance maxCooks
   # colnames(9): DRG_CS1_S4 DRG_CS2_S5 ... NG_F_S6 NG_F_S7
   # colData names(2): SampleNames condition
   #
    class((row.names(dds_drg_nod)))
   # [1] "character"

   ### our dds object includes ensemble ids as row.names stored in character
      ### we will need to prepare a version of this dds object having gene symbols as row.names.
      ### Because vsMAPLOT function takes row.names of dds object to highlight gene names on the graph.

     ### Retrieval of gene symbols for ensemble ids:

    BiocManager::install("org.Mm.eg.db")

    dds_drg_nodose_gene_symbols = mapIds(org.Mm.eg.db, keys = row.names(dds_drg_nod), column = 'SYMBOL', keytype = 'ENSEMBL')




    class(dds_drg_nodose_gene_symbols)
   # [1] "character"

   #### Store a version of original dds object:

   dds_drg_nodose_rownames_changed_w_symbols = dds_drg_nod
    ## Replace rownames with gene symbols:
    row.names(dds_drg_nodose_rownames_changed_w_symbols) = dds_drg_nodose_gene_symbols
    ### check if rownames are properly replaced:

     dds_drg_nodose_rownames_changed_w_symbols
   # class: DESeqDataSet
   # dim: 25202 9
   # metadata(1): version
   # assays(5): counts avgTxLength normalizationFactors mu
   # cooks
   # rownames(25202): Gnai3 Cdc45 ... NA NA
   # rowData names(29): baseMean baseVar ... deviance maxCooks
   # colnames(9): DRG_CS1_S4 DRG_CS2_S5 ... NG_F_S6 NG_F_S7
   # colData names(2): SampleNames condition
   #
   # ###

   # Check if gene symbols used as row.names contain any "NA":

      length(dds_drg_nodose_gene_symbols)
   # [1] 25202

   ## total number: 25502

     ### To view number of missing row.names:

      table(dds_drg_nodose_gene_symbols == "NA")

   # FALSE
   # 21011

   ## About 4000 are "NA".

   ### Change missing values with "no_name" character,

   ### ifelse (test, yes, no) function is used.
   ### if result of the test is yes, the value is replaced with
   # the input entered for "yes = paramater", if result is no,
   # it is replaced with the input enetered as "no= parameter".
   #
   ###

    dds_drg_nodose_gene_symbols_NA_replaced = ifelse(is.na(dds_drg_nodose_gene_symbols), "no_name", dds_drg_nodose_gene_symbols)

   ### we used is.na function wihich checks each residue in the provided
   # input and if missing value (NA) is encountered, replace it with "no_name".
   # If not encountered, leaves it unchanged. Note that dds_drg_nodose_gene_symbols
   # variable stores original gene names with NAs.

   ## Change the row.names of the dds object with NA replaced version:

    row.names(dds_drg_nodose_rownames_changed_w_symbols) = dds_drg_nodose_gene_symbols_NA_replaced

   ## Check if no_name is added:

    dds_drg_nodose_rownames_changed_w_symbols


   ### DONE.

   ###Now, MA plots will be drawn with select gene names higlighted on:

   ### DRG genes that will be highlighted on MA Plot are stored in rthe following variable:

     DRG_MA_plot_genes = c("Tnfaip8", "P2rx2" ,"Mrgprd", "Trpv1", "Chrna4","Trpm8","Scn10a","Scn11a","Il31ra","Mrgprx1","Mrgpra3" ,"Nppb")

   ### Same for nodose genes:

    Nodose_MA_plot_genes = c("P2rx2", "Ptprr", "Ptgdr", "Cckar", "Fgf3", "Scn10a", "Trpv1", "Trpm8")

   #### MA Plot function is applied with gene names supplied to the "highlight =" parameter:

   ### Nodose:

   vsMAPlot(x = 'cont_nodose', y = 'sample_nodose',data = dds_drg_nodose_rownames_changed_w_symbols, d.factor = 'condition', type = 'deseq',padj = 0.05, y.lim = NULL, lfc = 1.5, title = TRUE,legend = TRUE, grid = TRUE, highlight = Nodose_MA_plot_genes)

   ### DRG:

    vsMAPlot(x = 'cont_drg', y = 'sample_drg',data = dds_drg_nodose_rownames_changed_w_symbols, d.factor = 'condition', type = 'deseq',padj = 0.05, y.lim = NULL, lfc = 1.5, title = TRUE,legend = TRUE, grid = TRUE, highlight = DRG_MA_plot_genes)

   ### Saved as tiff and also US Legal PDF.

############################################################################################################

 ### Prepare a FPKM matrix containinig common significant genes for DRG and NG samples:


    ### we get the intersection of row.names from DRG significant and Nodose significat genes.

    both_DRG_and_Nodose_sig_genes = intersect(row.names(res_DRG_F_vs_cont_drg_na_omit_symbols_rm_dup_all_sig), row.names(res_NODS_F_vs_cont_nodose_na_omit_symbols_rm_dup_all_sig))
    length(both_DRG_and_Nodose_sig_genes)

    # [1] 349

    #### we filter our main matrix with these common genes' ensembl ids.

     fpkm_common_sig_genes_DRG_NOD = fpkm_drg_nod_matrix_w_padj[both_DRG_and_Nodose_sig_genes,]
     dim(fpkm_common_sig_genes_DRG_NOD)
    # [1] 349  13

    head(fpkm_common_sig_genes_DRG_NOD)


    ### we add the log2 fold change values from the DRG results data:

     fpkm_common_sig_genes_DRG_NOD$log2_FC_DRG = res_DRG_F_NA_keeped[row.names(fpkm_common_sig_genes_DRG_NOD),]$log2FoldChange

    ### we add the loh2 FC values from the nodose results data:

     fpkm_common_sig_genes_DRG_NOD$log2_FC_NG = res_NODS_F_vs_cont_nodose[row.names(fpkm_common_sig_genes_DRG_NOD),]$log2FoldChange
     head(fpkm_common_sig_genes_DRG_NOD)


    ### to have the matrix with only common UP genes, we firstly narrow down our matrix with genes up in DRG:

     fpkm_common_UP_sig_genes_DRG_NOD= fpkm_common_sig_genes_DRG_NOD[which(fpkm_common_sig_genes_DRG_NOD$log2_FC_DRG >0),]
     dim(fpkm_common_UP_sig_genes_DRG_NOD)
    # [1] 16 15

    ### In this new matrix of DRG up gens, we further shorlist our matrix with NG up genes:

     fpkm_common_UP_sig_genes_DRG_NOD= fpkm_common_UP_sig_genes_DRG_NOD[which(fpkm_common_UP_sig_genes_DRG_NOD$log2_FC_NG >0),]
     dim(fpkm_common_UP_sig_genes_DRG_NOD)
    # [1] 12 15

    ### Eventually, we obtained genes up in both DRG and NGs.

     fpkm_common_UP_sig_genes_DRG_NOD



    ### By the same logical approach, we obtain the matrix of down genes comon in both DRG and NG:

     fpkm_common_DOWN_sig_genes_DRG_NOD= fpkm_common_sig_genes_DRG_NOD[which(fpkm_common_sig_genes_DRG_NOD$log2_FC_DRG <0),]
     dim(fpkm_common_DOWN_sig_genes_DRG_NOD)
    # [1] 333  15
     fpkm_common_DOWN_sig_genes_DRG_NOD= fpkm_common_DOWN_sig_genes_DRG_NOD[which(fpkm_common_DOWN_sig_genes_DRG_NOD$log2_FC_NG <0),]
     dim(fpkm_common_DOWN_sig_genes_DRG_NOD)
    # [1] 333  15

    ### Finally, genes that show opposite expression patrerns in DRGs and NGs:

    fpkm_common_DRG_NOD_different_sig_genes= row.names(fpkm_common_sig_genes_DRG_NOD)

    ### Sequentially, we find the difference of all sig genes from the both DRG and NG DOWN genes:

    fpkm_common_DRG_NOD_different_sig_genes = setdiff(fpkm_common_DRG_NOD_different_sig_genes, row.names(fpkm_common_DOWN_sig_genes_DRG_NOD))

    ### Then, we find the difference of this new matrix from the genes being up in both DRG and NG:

    fpkm_common_DRG_NOD_different_sig_genes = setdiff(fpkm_common_DRG_NOD_different_sig_genes, row.names(fpkm_common_UP_sig_genes_DRG_NOD))

      length(fpkm_common_DRG_NOD_different_sig_genes)
    # [1] 4
     fpkm_common_DRG_NOD_different_sig_genes= fpkm_common_sig_genes_DRG_NOD[fpkm_common_DRG_NOD_different_sig_genes,]
     fpkm_common_DRG_NOD_different_sig_genes


    ### Length of data matrices were checked. Their summ equals to the initial fpkm common significant gene matrix.

    #### EXPORTS:

     write.csv(fpkm_common_DOWN_sig_genes_DRG_NOD, file = "fpkm_common_DOWN_sig_genes_DRG_NOD.txt")
     write.csv(fpkm_common_UP_sig_genes_DRG_NOD, file = "fpkm_common_UP_sig_genes_DRG_NOD.txt")
     write.csv(fpkm_common_DRG_NOD_different_sig_genes, file = "fpkm_oppositely_expressed_DRg_NOD_sig_genes.txt")

     sessionInfo()

